{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%capture\n",
    "# !pip install git+https://github.com/biothings/biothings_explorer#egg=biothings_explorer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import requests\n",
    "from urllib.request import urlopen\n",
    "from bs4 import BeautifulSoup\n",
    "from datetime import date\n",
    "import datetime as dt\n",
    "import pandas as pd\n",
    "import sys\n",
    "import csv\n",
    "import numpy as np\n",
    "import multiprocessing\n",
    "import re\n",
    "from collections import namedtuple\n",
    "# For pausing\n",
    "from time import sleep\n",
    "import findspark\n",
    "findspark.init()\n",
    "import pyspark\n",
    "os.environ['PYSPARK_SUBMIT_ARGS'] = \"\"\"--name job_name --master local --conf spark.dynamicAllocation.enabled=true pyspark-shell\"\"\"\n",
    "os.environ[\"JAVA_HOME\"] = \"/Library/Java/JavaVirtualMachines/jdk1.8.0_251.jdk/Contents/Home\"\n",
    "import pickle\n",
    "import inspect\n",
    "from itertools import repeat\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# adding Folder_2/subfolder to the system path\n",
    "sys.path.insert(0, '/Volumes/TOSHIBA_EXT/ISB/clinical_trials/pymetamap-master')\n",
    "from pymetamap import MetaMap\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import psycopg2\n",
    "import pandas.io.sql as sqlio\n",
    "# from tabulate import tabulate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import argparse\n",
    "import lxml\n",
    "import config\n",
    "import pickle\n",
    "\n",
    "# from Authentication import *\n",
    "import requests\n",
    "import json\n",
    "from requests.adapters import HTTPAdapter\n",
    "from requests.packages.urllib3.util.retry import Retry\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = \"https://aact.ctti-clinicaltrials.org/pipe_files\"\n",
    "response = requests.get(url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "soup = BeautifulSoup(response.text)\n",
    "# print(soup.prettify())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# https://aact.ctti-clinicaltrials.org/static/exported_files/monthly/20220301_pipe-delimited-export.zip\n",
    "# https://aact.ctti-clinicaltrials.org/static/exported_files/monthly/20220801_pipe-delimited-export.zip\n",
    "upload_dates = []\n",
    "zip_files = []\n",
    "links = []\n",
    "body = soup.find_all('td', attrs={'class': 'file-archive'}) #Find all\n",
    "for el in body:\n",
    "    tags = el.find('a')\n",
    "    try:\n",
    "        if 'href' in tags.attrs:   # looking for href inside anchor tag    \n",
    "            link = \"https://aact.ctti-clinicaltrials.org\" + tags.get('href')\n",
    "            links.append(link)\n",
    "            last_upload = link.split(\"/\")[-1]\n",
    "            zip_files.append(last_upload)\n",
    "            date_upload = last_upload.split(\"_\")[0]\n",
    "            upload_dates.append(date_upload)    # appending link to list of links\n",
    "    except:    # pass if list missing anchor tag or anchor tag does not has a href params \n",
    "        pass\n",
    "    \n",
    "print(upload_dates)\n",
    "\n",
    "upload_dates = [dt.datetime.strptime(date, '%Y%m%d').date() for date in upload_dates] # convert all strings in list into datetime objects\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "most_recent_date = max(upload_dates)\n",
    "# print(most_recent_date)\n",
    "\n",
    "today = dt.date.today()\n",
    "\n",
    "# UNCOMMENT LINE BELOW WHEN IN PRODUCTION, THIS IS FOR TESTING PURPOSES\n",
    "most_recent_date = min(upload_dates)\n",
    "\n",
    "# fix this to run entire script should the current date be \n",
    "if most_recent_date < today:\n",
    "#     date_file_df = pd.DataFrame(list(zip(upload_dates, zip_files, links)))\n",
    "#     print(links[0])\n",
    "    resp = urlopen(links[0])\n",
    "else:\n",
    "    print(\"Local instance of clinical trials data is already up to date.\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', 3000)\n",
    "# date_file_df[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.getcwd()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# connect to DB and get the column names of the table\n",
    "con = None\n",
    "con = psycopg2.connect(database=\"aact\")\n",
    "con.rollback()\n",
    "cursor = con.cursor()\n",
    "\n",
    "con.autocommit = True # SQL statement is treated as a transaction and is automatically committed right after it is executed\n",
    "# grab the conditions\n",
    "sql = '''SELECT * FROM ctgov.conditions;'''\n",
    "cursor.execute(sql)\n",
    "column_names = [desc[0] for desc in cursor.description]\n",
    "tuples = cursor.fetchall()\n",
    "conditions_df = pd.DataFrame(tuples, columns=column_names)\n",
    "\n",
    "#grab the browse_conditions\n",
    "sql = '''SELECT * FROM ctgov.browse_conditions;'''\n",
    "cursor.execute(sql)\n",
    "column_names = [desc[0] for desc in cursor.description]\n",
    "tuples = cursor.fetchall()\n",
    "browse_conditions_df = pd.DataFrame(tuples, columns=column_names)\n",
    "\n",
    "#grab the interventions\n",
    "sql = '''SELECT * FROM ctgov.interventions;'''\n",
    "cursor.execute(sql)\n",
    "column_names = [desc[0] for desc in cursor.description]\n",
    "tuples = cursor.fetchall()\n",
    "interventions_df = pd.DataFrame(tuples, columns=column_names)\n",
    "\n",
    "#grab the browse_interventions\n",
    "sql = '''SELECT * FROM ctgov.browse_interventions;'''\n",
    "cursor.execute(sql)\n",
    "column_names = [desc[0] for desc in cursor.description]\n",
    "tuples = cursor.fetchall()\n",
    "browse_interventions_df = pd.DataFrame(tuples, columns=column_names)\n",
    "\n",
    "con.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>int_id</th>\n",
       "      <th>int_nctid</th>\n",
       "      <th>int_type</th>\n",
       "      <th>int_name</th>\n",
       "      <th>int_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4547568</td>\n",
       "      <td>NCT01340339</td>\n",
       "      <td>Device</td>\n",
       "      <td>Phototherapy</td>\n",
       "      <td>Fluorescent reverse phototherapy (7 white ligh...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4547569</td>\n",
       "      <td>NCT01340339</td>\n",
       "      <td>Device</td>\n",
       "      <td>Phototherapy</td>\n",
       "      <td>super LED reverse phototherapy (17 bulbs arran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4547570</td>\n",
       "      <td>NCT01340365</td>\n",
       "      <td>Behavioral</td>\n",
       "      <td>Tai Chi Exercise</td>\n",
       "      <td>Practicing Tai Chi exercise 4 times a week for...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4547571</td>\n",
       "      <td>NCT01340365</td>\n",
       "      <td>Behavioral</td>\n",
       "      <td>Tai Chi</td>\n",
       "      <td>Usual care, individuals attend testing session...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4547572</td>\n",
       "      <td>NCT01340391</td>\n",
       "      <td>Device</td>\n",
       "      <td>Omnicast</td>\n",
       "      <td>dorsal splint 2-5 weeks</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685271</th>\n",
       "      <td>4139681</td>\n",
       "      <td>NCT03104712</td>\n",
       "      <td>Other</td>\n",
       "      <td>Rice + Pesto</td>\n",
       "      <td>50g available carbohydrate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685272</th>\n",
       "      <td>4139682</td>\n",
       "      <td>NCT03104465</td>\n",
       "      <td>Behavioral</td>\n",
       "      <td>Mindfulness Training</td>\n",
       "      <td>interactive. web-based mindfulness training co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685273</th>\n",
       "      <td>4139683</td>\n",
       "      <td>NCT03104361</td>\n",
       "      <td>Biological</td>\n",
       "      <td>Platelet-Rich Plasma</td>\n",
       "      <td>Patients who meet all eligible requirements fo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685274</th>\n",
       "      <td>4139684</td>\n",
       "      <td>NCT03104283</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Apatinib</td>\n",
       "      <td>take apatinib orally until disease progression...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685275</th>\n",
       "      <td>4139685</td>\n",
       "      <td>NCT03104231</td>\n",
       "      <td>Other</td>\n",
       "      <td>three-dimensional (3D) movie</td>\n",
       "      <td>Brief VR-3D will be shown using the oculus gla...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>685276 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         int_id    int_nctid    int_type                      int_name  \\\n",
       "0       4547568  NCT01340339      Device                  Phototherapy   \n",
       "1       4547569  NCT01340339      Device                  Phototherapy   \n",
       "2       4547570  NCT01340365  Behavioral              Tai Chi Exercise   \n",
       "3       4547571  NCT01340365  Behavioral                       Tai Chi   \n",
       "4       4547572  NCT01340391      Device                      Omnicast   \n",
       "...         ...          ...         ...                           ...   \n",
       "685271  4139681  NCT03104712       Other                  Rice + Pesto   \n",
       "685272  4139682  NCT03104465  Behavioral          Mindfulness Training   \n",
       "685273  4139683  NCT03104361  Biological          Platelet-Rich Plasma   \n",
       "685274  4139684  NCT03104283        Drug                      Apatinib   \n",
       "685275  4139685  NCT03104231       Other  three-dimensional (3D) movie   \n",
       "\n",
       "                                          int_description  \n",
       "0       Fluorescent reverse phototherapy (7 white ligh...  \n",
       "1       super LED reverse phototherapy (17 bulbs arran...  \n",
       "2       Practicing Tai Chi exercise 4 times a week for...  \n",
       "3       Usual care, individuals attend testing session...  \n",
       "4                                 dorsal splint 2-5 weeks  \n",
       "...                                                   ...  \n",
       "685271                         50g available carbohydrate  \n",
       "685272  interactive. web-based mindfulness training co...  \n",
       "685273  Patients who meet all eligible requirements fo...  \n",
       "685274  take apatinib orally until disease progression...  \n",
       "685275  Brief VR-3D will be shown using the oculus gla...  \n",
       "\n",
       "[685276 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interventions_df = interventions_df.rename(columns={'id': 'int_id',\n",
    "                                                    'nct_id': 'int_nctid',\n",
    "                                                    'intervention_type': 'int_type',\n",
    "                                                    'name': 'int_name',\n",
    "                                                    'description': 'int_description'})\n",
    "interventions_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "interventions_df = interventions_df.drop(columns=['int_id', 'int_description'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>con_id</th>\n",
       "      <th>con_nctid</th>\n",
       "      <th>con_name</th>\n",
       "      <th>con_downcase_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3072834</td>\n",
       "      <td>NCT03254264</td>\n",
       "      <td>Autism Spectrum Disorder</td>\n",
       "      <td>autism spectrum disorder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3072835</td>\n",
       "      <td>NCT03254329</td>\n",
       "      <td>Human Milk Nutrient Reference Values</td>\n",
       "      <td>human milk nutrient reference values</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3349485</td>\n",
       "      <td>NCT03460652</td>\n",
       "      <td>ADHD</td>\n",
       "      <td>adhd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3349486</td>\n",
       "      <td>NCT03460899</td>\n",
       "      <td>Diabetes Mellitus With Hypoglycemia</td>\n",
       "      <td>diabetes mellitus with hypoglycemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3072836</td>\n",
       "      <td>NCT03254342</td>\n",
       "      <td>Major Depressive Disorder</td>\n",
       "      <td>major depressive disorder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688783</th>\n",
       "      <td>3072052</td>\n",
       "      <td>NCT03261687</td>\n",
       "      <td>Pregnancy Related</td>\n",
       "      <td>pregnancy related</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688784</th>\n",
       "      <td>3072053</td>\n",
       "      <td>NCT03261687</td>\n",
       "      <td>Pelvic Girdle Pain</td>\n",
       "      <td>pelvic girdle pain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688785</th>\n",
       "      <td>3072054</td>\n",
       "      <td>NCT03261622</td>\n",
       "      <td>Faecal Incontinence</td>\n",
       "      <td>faecal incontinence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688786</th>\n",
       "      <td>3072055</td>\n",
       "      <td>NCT03261622</td>\n",
       "      <td>Fecal Incontinence</td>\n",
       "      <td>fecal incontinence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688787</th>\n",
       "      <td>3072056</td>\n",
       "      <td>NCT03261934</td>\n",
       "      <td>Risk Assessment of Chronic Atrophic Gastritis ...</td>\n",
       "      <td>risk assessment of chronic atrophic gastritis ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>688788 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         con_id    con_nctid  \\\n",
       "0       3072834  NCT03254264   \n",
       "1       3072835  NCT03254329   \n",
       "2       3349485  NCT03460652   \n",
       "3       3349486  NCT03460899   \n",
       "4       3072836  NCT03254342   \n",
       "...         ...          ...   \n",
       "688783  3072052  NCT03261687   \n",
       "688784  3072053  NCT03261687   \n",
       "688785  3072054  NCT03261622   \n",
       "688786  3072055  NCT03261622   \n",
       "688787  3072056  NCT03261934   \n",
       "\n",
       "                                                 con_name  \\\n",
       "0                                Autism Spectrum Disorder   \n",
       "1                    Human Milk Nutrient Reference Values   \n",
       "2                                                    ADHD   \n",
       "3                     Diabetes Mellitus With Hypoglycemia   \n",
       "4                               Major Depressive Disorder   \n",
       "...                                                   ...   \n",
       "688783                                  Pregnancy Related   \n",
       "688784                                 Pelvic Girdle Pain   \n",
       "688785                                Faecal Incontinence   \n",
       "688786                                 Fecal Incontinence   \n",
       "688787  Risk Assessment of Chronic Atrophic Gastritis ...   \n",
       "\n",
       "                                        con_downcase_name  \n",
       "0                                autism spectrum disorder  \n",
       "1                    human milk nutrient reference values  \n",
       "2                                                    adhd  \n",
       "3                     diabetes mellitus with hypoglycemia  \n",
       "4                               major depressive disorder  \n",
       "...                                                   ...  \n",
       "688783                                  pregnancy related  \n",
       "688784                                 pelvic girdle pain  \n",
       "688785                                faecal incontinence  \n",
       "688786                                 fecal incontinence  \n",
       "688787  risk assessment of chronic atrophic gastritis ...  \n",
       "\n",
       "[688788 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conditions_df = conditions_df.rename(columns={'id': 'con_id',\n",
    "                                              'nct_id': 'con_nctid',\n",
    "                                              'name': 'con_name',\n",
    "                                              'downcase_name': 'con_downcase_name'})\n",
    "conditions_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "conditions_df = conditions_df.drop(columns=['con_id', 'con_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>browseint_id</th>\n",
       "      <th>browseint_nctid</th>\n",
       "      <th>browseint_meshterm</th>\n",
       "      <th>browseint_meshterm_downcase</th>\n",
       "      <th>browseint_meshtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4935968</td>\n",
       "      <td>NCT04327843</td>\n",
       "      <td>Haloperidol</td>\n",
       "      <td>haloperidol</td>\n",
       "      <td>mesh-list</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4935969</td>\n",
       "      <td>NCT04327843</td>\n",
       "      <td>Haloperidol decanoate</td>\n",
       "      <td>haloperidol decanoate</td>\n",
       "      <td>mesh-list</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4935970</td>\n",
       "      <td>NCT04327843</td>\n",
       "      <td>Antiemetics</td>\n",
       "      <td>antiemetics</td>\n",
       "      <td>mesh-ancestor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5198960</td>\n",
       "      <td>NCT00808223</td>\n",
       "      <td>Alefacept</td>\n",
       "      <td>alefacept</td>\n",
       "      <td>mesh-list</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5198961</td>\n",
       "      <td>NCT00808223</td>\n",
       "      <td>Dermatologic Agents</td>\n",
       "      <td>dermatologic agents</td>\n",
       "      <td>mesh-ancestor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335262</th>\n",
       "      <td>4400668</td>\n",
       "      <td>NCT03100786</td>\n",
       "      <td>Antineoplastic Agents, Hormonal</td>\n",
       "      <td>antineoplastic agents, hormonal</td>\n",
       "      <td>mesh-ancestor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335263</th>\n",
       "      <td>4400669</td>\n",
       "      <td>NCT03100786</td>\n",
       "      <td>Antineoplastic Agents</td>\n",
       "      <td>antineoplastic agents</td>\n",
       "      <td>mesh-ancestor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335264</th>\n",
       "      <td>4400670</td>\n",
       "      <td>NCT03100565</td>\n",
       "      <td>Lidocaine</td>\n",
       "      <td>lidocaine</td>\n",
       "      <td>mesh-list</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335265</th>\n",
       "      <td>4400671</td>\n",
       "      <td>NCT03100565</td>\n",
       "      <td>Anesthetics, Local</td>\n",
       "      <td>anesthetics, local</td>\n",
       "      <td>mesh-ancestor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335266</th>\n",
       "      <td>4400672</td>\n",
       "      <td>NCT03100565</td>\n",
       "      <td>Anesthetics</td>\n",
       "      <td>anesthetics</td>\n",
       "      <td>mesh-ancestor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1335267 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         browseint_id browseint_nctid               browseint_meshterm  \\\n",
       "0             4935968     NCT04327843                      Haloperidol   \n",
       "1             4935969     NCT04327843            Haloperidol decanoate   \n",
       "2             4935970     NCT04327843                      Antiemetics   \n",
       "3             5198960     NCT00808223                        Alefacept   \n",
       "4             5198961     NCT00808223              Dermatologic Agents   \n",
       "...               ...             ...                              ...   \n",
       "1335262       4400668     NCT03100786  Antineoplastic Agents, Hormonal   \n",
       "1335263       4400669     NCT03100786            Antineoplastic Agents   \n",
       "1335264       4400670     NCT03100565                        Lidocaine   \n",
       "1335265       4400671     NCT03100565               Anesthetics, Local   \n",
       "1335266       4400672     NCT03100565                      Anesthetics   \n",
       "\n",
       "             browseint_meshterm_downcase browseint_meshtype  \n",
       "0                            haloperidol          mesh-list  \n",
       "1                  haloperidol decanoate          mesh-list  \n",
       "2                            antiemetics      mesh-ancestor  \n",
       "3                              alefacept          mesh-list  \n",
       "4                    dermatologic agents      mesh-ancestor  \n",
       "...                                  ...                ...  \n",
       "1335262  antineoplastic agents, hormonal      mesh-ancestor  \n",
       "1335263            antineoplastic agents      mesh-ancestor  \n",
       "1335264                        lidocaine          mesh-list  \n",
       "1335265               anesthetics, local      mesh-ancestor  \n",
       "1335266                      anesthetics      mesh-ancestor  \n",
       "\n",
       "[1335267 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "browse_interventions_df = browse_interventions_df.rename(columns={'id': 'browseint_id',\n",
    "                                                                  'nct_id': 'browseint_nctid',\n",
    "                                                                  'mesh_term': 'browseint_meshterm',\n",
    "                                                                  'downcase_mesh_term': 'browseint_meshterm_downcase',\n",
    "                                                                  'mesh_type': 'browseint_meshtype'})\n",
    "browse_interventions_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "browse_interventions_df = browse_interventions_df.drop(columns=['browseint_id', 'browseint_meshterm'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>browsecon_id</th>\n",
       "      <th>browsecon_nctid</th>\n",
       "      <th>browsecon_meshterm</th>\n",
       "      <th>browsecon_meshterm_downcase</th>\n",
       "      <th>browsecon_meshtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9631397</td>\n",
       "      <td>NCT04035564</td>\n",
       "      <td>Hyponatremia</td>\n",
       "      <td>hyponatremia</td>\n",
       "      <td>mesh-list</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9631398</td>\n",
       "      <td>NCT04035564</td>\n",
       "      <td>Water-Electrolyte Imbalance</td>\n",
       "      <td>water-electrolyte imbalance</td>\n",
       "      <td>mesh-ancestor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9631399</td>\n",
       "      <td>NCT04035564</td>\n",
       "      <td>Metabolic Diseases</td>\n",
       "      <td>metabolic diseases</td>\n",
       "      <td>mesh-ancestor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9631400</td>\n",
       "      <td>NCT04032652</td>\n",
       "      <td>Colitis</td>\n",
       "      <td>colitis</td>\n",
       "      <td>mesh-list</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9631401</td>\n",
       "      <td>NCT04032652</td>\n",
       "      <td>Colitis, Ulcerative</td>\n",
       "      <td>colitis, ulcerative</td>\n",
       "      <td>mesh-list</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573556</th>\n",
       "      <td>8601072</td>\n",
       "      <td>NCT03213899</td>\n",
       "      <td>Disease Attributes</td>\n",
       "      <td>disease attributes</td>\n",
       "      <td>mesh-ancestor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573557</th>\n",
       "      <td>8601073</td>\n",
       "      <td>NCT03213899</td>\n",
       "      <td>Pathologic Processes</td>\n",
       "      <td>pathologic processes</td>\n",
       "      <td>mesh-ancestor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573558</th>\n",
       "      <td>8601074</td>\n",
       "      <td>NCT03214016</td>\n",
       "      <td>Hypertension</td>\n",
       "      <td>hypertension</td>\n",
       "      <td>mesh-list</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573559</th>\n",
       "      <td>8601075</td>\n",
       "      <td>NCT03214016</td>\n",
       "      <td>Vascular Diseases</td>\n",
       "      <td>vascular diseases</td>\n",
       "      <td>mesh-ancestor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2573560</th>\n",
       "      <td>8601076</td>\n",
       "      <td>NCT03214016</td>\n",
       "      <td>Cardiovascular Diseases</td>\n",
       "      <td>cardiovascular diseases</td>\n",
       "      <td>mesh-ancestor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2573561 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         browsecon_id browsecon_nctid           browsecon_meshterm  \\\n",
       "0             9631397     NCT04035564                 Hyponatremia   \n",
       "1             9631398     NCT04035564  Water-Electrolyte Imbalance   \n",
       "2             9631399     NCT04035564           Metabolic Diseases   \n",
       "3             9631400     NCT04032652                      Colitis   \n",
       "4             9631401     NCT04032652          Colitis, Ulcerative   \n",
       "...               ...             ...                          ...   \n",
       "2573556       8601072     NCT03213899           Disease Attributes   \n",
       "2573557       8601073     NCT03213899         Pathologic Processes   \n",
       "2573558       8601074     NCT03214016                 Hypertension   \n",
       "2573559       8601075     NCT03214016            Vascular Diseases   \n",
       "2573560       8601076     NCT03214016      Cardiovascular Diseases   \n",
       "\n",
       "         browsecon_meshterm_downcase browsecon_meshtype  \n",
       "0                       hyponatremia          mesh-list  \n",
       "1        water-electrolyte imbalance      mesh-ancestor  \n",
       "2                 metabolic diseases      mesh-ancestor  \n",
       "3                            colitis          mesh-list  \n",
       "4                colitis, ulcerative          mesh-list  \n",
       "...                              ...                ...  \n",
       "2573556           disease attributes      mesh-ancestor  \n",
       "2573557         pathologic processes      mesh-ancestor  \n",
       "2573558                 hypertension          mesh-list  \n",
       "2573559            vascular diseases      mesh-ancestor  \n",
       "2573560      cardiovascular diseases      mesh-ancestor  \n",
       "\n",
       "[2573561 rows x 5 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "browse_conditions_df = browse_conditions_df.rename(columns={'id': 'browsecon_id',\n",
    "                                                            'nct_id': 'browsecon_nctid',\n",
    "                                                            'mesh_term': 'browsecon_meshterm',\n",
    "                                                            'downcase_mesh_term': 'browsecon_meshterm_downcase',\n",
    "                                                            'mesh_type': 'browsecon_meshtype'})\n",
    "browse_conditions_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "browse_conditions_df = browse_conditions_df.drop(columns=['browsecon_id', 'browsecon_meshterm'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "350361\n",
      "95250\n",
      "4280\n",
      "3744\n"
     ]
    }
   ],
   "source": [
    "print(len(interventions_df['int_name'].unique()))\n",
    "print(len(conditions_df['con_downcase_name'].unique()))\n",
    "print(len(browse_conditions_df['browsecon_meshterm_downcase'].unique()))\n",
    "print(len(browse_interventions_df['browseint_meshterm_downcase'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>con_nctid</th>\n",
       "      <th>con_downcase_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCT03254264</td>\n",
       "      <td>autism spectrum disorder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCT03254329</td>\n",
       "      <td>human milk nutrient reference values</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCT03460652</td>\n",
       "      <td>adhd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCT03460899</td>\n",
       "      <td>diabetes mellitus with hypoglycemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCT03254342</td>\n",
       "      <td>major depressive disorder</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688783</th>\n",
       "      <td>NCT03261687</td>\n",
       "      <td>pregnancy related</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688784</th>\n",
       "      <td>NCT03261687</td>\n",
       "      <td>pelvic girdle pain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688785</th>\n",
       "      <td>NCT03261622</td>\n",
       "      <td>faecal incontinence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688786</th>\n",
       "      <td>NCT03261622</td>\n",
       "      <td>fecal incontinence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688787</th>\n",
       "      <td>NCT03261934</td>\n",
       "      <td>risk assessment of chronic atrophic gastritis ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>688788 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          con_nctid                                  con_downcase_name\n",
       "0       NCT03254264                           autism spectrum disorder\n",
       "1       NCT03254329               human milk nutrient reference values\n",
       "2       NCT03460652                                               adhd\n",
       "3       NCT03460899                diabetes mellitus with hypoglycemia\n",
       "4       NCT03254342                          major depressive disorder\n",
       "...             ...                                                ...\n",
       "688783  NCT03261687                                  pregnancy related\n",
       "688784  NCT03261687                                 pelvic girdle pain\n",
       "688785  NCT03261622                                faecal incontinence\n",
       "688786  NCT03261622                                 fecal incontinence\n",
       "688787  NCT03261934  risk assessment of chronic atrophic gastritis ...\n",
       "\n",
       "[688788 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conditions_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.merge(conditions_df, interventions_df, left_on='con_nctid', right_on = 'int_nctid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1256248, 5)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dedup = df.drop_duplicates(subset = ['con_downcase_name', 'int_name'],\n",
    "                                      keep = 'first').reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dedup.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with pd.option_context('display.max_rows', None, 'display.max_columns', None):  # more options can be specified also\n",
    "    display(df_dedup.head(1000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dedup[['con_downcase_name', 'int_name']].to_csv('ClinTrials_KG_edges.csv', sep ='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# notes on creating edge file\n",
    "\n",
    "go to /proj/arivale/gglusman/KGs\n",
    "\n",
    "once there look into wellness/source/wellness_kg_edges_v1.6.tsv\n",
    "\n",
    "notice the first line: subject\tpredicate\tobject\trelation\tsubject_name\tobject_name\tcategory\tN\tType_of_relationship\tStrength_of_relationship\tqualifiers\tqualifier_value\tBonferroni_pval\n",
    "\n",
    "subject, predicate, object are required (maybe some others, don’t remember)\n",
    "\n",
    "…and they should be CURIEs\n",
    "\n",
    "once we map concepts to biolink, mesh etc we’ll have curies\n",
    "\n",
    "meanwhile, the content you have fits subject_name, object_name\n",
    "\n",
    "the predicate is what we made up… let’s say ‘evaluated_for’\n",
    "\n",
    "for short\n",
    "\n",
    "we’ll have to find a suitable biolink predicate, or suggest a new one\n",
    "\n",
    "so interventions would be subjects, and conditions would be objects\n",
    "\n",
    "looking at the several first lines in your file, I see some that make sense, others that seem to suggest there are data problems\n",
    "\n",
    "for example, “metastatic lung cancer\tPathways” -> Pathways doesn’t seem to be an intervention\n",
    "\n",
    "‘placebo’ we’ll have to suppress\n",
    "\n",
    "make sure to include the NCT id in a (non-required) column\n",
    "\n",
    "it’s essentially part of the provenance\n",
    "\n",
    "we’ll want to add multiple additional columns with things like FDA approval status, sample size, etc etc etc\n",
    "\n",
    "look also into EHR/source/ehr_risk_kg_edges_2021_05_07.tsv\n",
    "\n",
    "…and of course replace ‘edges’ -> ‘nodes’ for their cognate nodes files\n",
    "\n",
    "wellness/source/wellness_kg_edges_v1.6.tsv -> wellness/source/wellness_kg_nodes_v1.6.tsv\n",
    "\n",
    "drop \"control\"\n",
    "\n",
    "‘control’ is clear, ‘control group’ too… but ‘Control Ostomy Barrier’ is likely not a ‘control’\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# My own notes\n",
    "\n",
    "Use pool/multiprocess to make faster\n",
    "resolve multiple possible CURIEs per concept"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# first get only relevant columns from DB\n",
    "ct_extract = pd.DataFrame(df_dedup[['con_nctid', 'con_downcase_name', 'int_type', 'int_name']])\n",
    "ct_extract = ct_extract.rename(columns={'con_nctid': 'nctid'})\n",
    "# get CURIE column for nct_id column (https://bioregistry.io/registry/clinicaltrials)\n",
    "\n",
    "ct_extract['nctid_curie'] = ct_extract['nctid']\n",
    "ct_extract['nctid_curie'] = 'clinicaltrials:' + ct_extract['nctid'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nctid</th>\n",
       "      <th>con_downcase_name</th>\n",
       "      <th>int_type</th>\n",
       "      <th>int_name</th>\n",
       "      <th>nctid_curie</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NCT03254264</td>\n",
       "      <td>autism spectrum disorder</td>\n",
       "      <td>Behavioral</td>\n",
       "      <td>Early Start Denver Model</td>\n",
       "      <td>clinicaltrials:NCT03254264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCT03254264</td>\n",
       "      <td>autism spectrum disorder</td>\n",
       "      <td>Behavioral</td>\n",
       "      <td>'as-usual' intervention</td>\n",
       "      <td>clinicaltrials:NCT03254264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NCT03254329</td>\n",
       "      <td>human milk nutrient reference values</td>\n",
       "      <td>Other</td>\n",
       "      <td>Assessment of human milk nutrient composition</td>\n",
       "      <td>clinicaltrials:NCT03254329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCT03460652</td>\n",
       "      <td>adhd</td>\n",
       "      <td>Drug</td>\n",
       "      <td>KP415 oral capsule</td>\n",
       "      <td>clinicaltrials:NCT03460652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCT03460899</td>\n",
       "      <td>diabetes mellitus with hypoglycemia</td>\n",
       "      <td>Other</td>\n",
       "      <td>Euglycaemic Clamp</td>\n",
       "      <td>clinicaltrials:NCT03460899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972680</th>\n",
       "      <td>NCT03261687</td>\n",
       "      <td>pelvic girdle pain</td>\n",
       "      <td>Other</td>\n",
       "      <td>water based exercise</td>\n",
       "      <td>clinicaltrials:NCT03261687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972681</th>\n",
       "      <td>NCT03261687</td>\n",
       "      <td>pelvic girdle pain</td>\n",
       "      <td>Other</td>\n",
       "      <td>land based exercise</td>\n",
       "      <td>clinicaltrials:NCT03261687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972682</th>\n",
       "      <td>NCT03261622</td>\n",
       "      <td>faecal incontinence</td>\n",
       "      <td>Device</td>\n",
       "      <td>Sacral nerve stimulation at different stimulat...</td>\n",
       "      <td>clinicaltrials:NCT03261622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972683</th>\n",
       "      <td>NCT03261622</td>\n",
       "      <td>fecal incontinence</td>\n",
       "      <td>Device</td>\n",
       "      <td>Sacral nerve stimulation at different stimulat...</td>\n",
       "      <td>clinicaltrials:NCT03261622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972684</th>\n",
       "      <td>NCT03261934</td>\n",
       "      <td>risk assessment of chronic atrophic gastritis ...</td>\n",
       "      <td>Other</td>\n",
       "      <td>No intervention</td>\n",
       "      <td>clinicaltrials:NCT03261934</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>972685 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              nctid                                  con_downcase_name  \\\n",
       "0       NCT03254264                           autism spectrum disorder   \n",
       "1       NCT03254264                           autism spectrum disorder   \n",
       "2       NCT03254329               human milk nutrient reference values   \n",
       "3       NCT03460652                                               adhd   \n",
       "4       NCT03460899                diabetes mellitus with hypoglycemia   \n",
       "...             ...                                                ...   \n",
       "972680  NCT03261687                                 pelvic girdle pain   \n",
       "972681  NCT03261687                                 pelvic girdle pain   \n",
       "972682  NCT03261622                                faecal incontinence   \n",
       "972683  NCT03261622                                 fecal incontinence   \n",
       "972684  NCT03261934  risk assessment of chronic atrophic gastritis ...   \n",
       "\n",
       "          int_type                                           int_name  \\\n",
       "0       Behavioral                           Early Start Denver Model   \n",
       "1       Behavioral                            'as-usual' intervention   \n",
       "2            Other      Assessment of human milk nutrient composition   \n",
       "3             Drug                                 KP415 oral capsule   \n",
       "4            Other                                  Euglycaemic Clamp   \n",
       "...            ...                                                ...   \n",
       "972680       Other                               water based exercise   \n",
       "972681       Other                                land based exercise   \n",
       "972682      Device  Sacral nerve stimulation at different stimulat...   \n",
       "972683      Device  Sacral nerve stimulation at different stimulat...   \n",
       "972684       Other                                    No intervention   \n",
       "\n",
       "                       nctid_curie  \n",
       "0       clinicaltrials:NCT03254264  \n",
       "1       clinicaltrials:NCT03254264  \n",
       "2       clinicaltrials:NCT03254329  \n",
       "3       clinicaltrials:NCT03460652  \n",
       "4       clinicaltrials:NCT03460899  \n",
       "...                            ...  \n",
       "972680  clinicaltrials:NCT03261687  \n",
       "972681  clinicaltrials:NCT03261687  \n",
       "972682  clinicaltrials:NCT03261622  \n",
       "972683  clinicaltrials:NCT03261622  \n",
       "972684  clinicaltrials:NCT03261934  \n",
       "\n",
       "[972685 rows x 5 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ct_extract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's map all the conditions to BTE\n",
    "# NOTE: we are attempting to match to BTE terms, but many of the conditions terms are not standardized and are misspelled\n",
    "# First map as many terms as possible to BTE. Whatever is left over, will use other ontologies to create CURIEs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# recommedation to first format the output as close to sample as possible"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "ct_final = pd.DataFrame(columns=['subject','predicate','object','relation','subject_name','object_name','category'])\n",
    "# no_col_names_df.columns = col_names_df.columns\n",
    "# chosen predicate = \"related_to\", bc \"associated_with\" implies a statistical relationship \n",
    "ct_final['subject'] = 'condition:' + ct_extract['con_downcase_name'].astype(str)\n",
    "ct_final['predicate'] = 'biolink:related_to biolink:clinically_demonstrated_to biolink: hypothesized_to'\n",
    "ct_final['object'] = 'intervention:' + ct_extract['int_name'].astype(str)   # this will not all be RxNorm CURIEs since some interventions are not drugs\n",
    "ct_final['relation'] = 'RO:????'\n",
    "ct_final.subject_name = ct_extract.con_downcase_name\n",
    "ct_final.object_name = ct_extract.int_name\n",
    "ct_final.category = 'biolink:Association'\n",
    "ct_final['nctid'] = ct_extract['nctid']\n",
    "ct_final['nctid_curie'] = ct_extract['nctid_curie']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>predicate</th>\n",
       "      <th>object</th>\n",
       "      <th>relation</th>\n",
       "      <th>subject_name</th>\n",
       "      <th>object_name</th>\n",
       "      <th>category</th>\n",
       "      <th>nctid</th>\n",
       "      <th>nctid_curie</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>condition:autism spectrum disorder</td>\n",
       "      <td>biolink:related_to biolink:clinically_demonstr...</td>\n",
       "      <td>intervention:Early Start Denver Model</td>\n",
       "      <td>RO:????</td>\n",
       "      <td>autism spectrum disorder</td>\n",
       "      <td>Early Start Denver Model</td>\n",
       "      <td>biolink:Association</td>\n",
       "      <td>NCT03254264</td>\n",
       "      <td>clinicaltrials:NCT03254264</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              subject  \\\n",
       "0  condition:autism spectrum disorder   \n",
       "\n",
       "                                           predicate  \\\n",
       "0  biolink:related_to biolink:clinically_demonstr...   \n",
       "\n",
       "                                  object relation              subject_name  \\\n",
       "0  intervention:Early Start Denver Model  RO:????  autism spectrum disorder   \n",
       "\n",
       "                object_name             category        nctid  \\\n",
       "0  Early Start Denver Model  biolink:Association  NCT03254264   \n",
       "\n",
       "                  nctid_curie  \n",
       "0  clinicaltrials:NCT03254264  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ct_final[:1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# interventions will have to be mapped to more than 1 vocabulary unlike conditions, which seems more easily mapped to SNOMED\n",
    "# first look at the types of interventions (unique on int_type in ct_extract dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/Kamileh/Work/ISB/NCATS_BiomedicalTranslator/Projects/ClinicalTrials/ETL_Python\n",
      "../../../../../../../../Volumes/TOSHIBA_EXT/ISB/clinical_trials/public_mm/bin\n"
     ]
    }
   ],
   "source": [
    "# get the directory of the the module\n",
    "print(os.getcwd())\n",
    "# Path\n",
    "path = \"/Volumes/TOSHIBA_EXT/ISB/clinical_trials/public_mm/bin\"  \n",
    "# Path of Start directory\n",
    "start = os.getcwd()\n",
    "print(os.path.relpath(\"/Volumes/TOSHIBA_EXT/ISB/clinical_trials/public_mm/bin\", os.getcwd()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# For pausing\n",
    "# from time import sleep\n",
    "\n",
    "# NOTE: YOU MUST START the SKR/Medpost Part-of-Speech Tagger Server and the Word Sense Disambiguation (WSD) Server IN TERMINAL\n",
    "# Setup UMLS Server\n",
    "metamap_base_dir = '/Volumes/TOSHIBA_EXT/ISB/clinical_trials/public_mm/'\n",
    "metamap_bin_dir = 'bin/metamap18'\n",
    "metamap_pos_server_dir = 'bin/skrmedpostctl'\n",
    "metamap_wsd_server_dir = 'bin/wsdserverctl'\n",
    "\n",
    "\n",
    "# Start servers\n",
    "os.system(metamap_base_dir + metamap_pos_server_dir + ' start') # Part of speech tagger\n",
    "os.system(metamap_base_dir + metamap_wsd_server_dir + ' start') # Word sense disambiguation \n",
    "\n",
    "# Sleep a bit to give time for these servers to start up\n",
    "sleep(60)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "mm = MetaMap.get_instance(metamap_base_dir + metamap_bin_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ConceptMMI(index='USER', mm='MMI', score='13.22', preferred_name='Myocardial Infarction', cui='C0027051', semtypes='[dsyn]', trigger='[\"Infarction, Myocardial\"-tx-1-\"myocardial infarction\"-noun-0]', location='TX', pos_info='4/21', tree_codes='C14.280.647.500;C14.907.585.500')\n",
      "\n",
      "\n",
      "ConceptMMI(index='USER', mm='MMI', score='3.75', preferred_name='Electrocardiogram: myocardial infarction (finding)', cui='C0428953', semtypes='[lbtr]', trigger='[\"MYOCARDIAL INFARCTION\"-tx-1-\"myocardial infarction\"-noun-0]', location='TX', pos_info='4/21', tree_codes='')\n",
      "\n",
      "\n",
      "ConceptMMI(index='USER', mm='MMI', score='3.75', preferred_name='Myocardial Infarction ECG Assessment', cui='C3810814', semtypes='[diap]', trigger='[\"Myocardial Infarction\"-tx-1-\"myocardial infarction\"-noun-0]', location='TX', pos_info='4/21', tree_codes='')\n",
      "\n",
      "\n",
      "ConceptMMI(index='USER', mm='MMI', score='3.75', preferred_name='Myocardial Infarction, CTCAE', cui='C4552959', semtypes='[fndg]', trigger='[\"Myocardial infarction\"-tx-1-\"myocardial infarction\"-noun-0]', location='TX', pos_info='4/21', tree_codes='')\n",
      "\n",
      "\n",
      "ConceptMMI(index='USER', mm='MMI', score='3.75', preferred_name='Myocardial infarction:Finding:Point in time:^Patient:Ordinal', cui='C2926063', semtypes='[clna]', trigger='[\"Myocardial infarction\"-tx-1-\"myocardial infarction\"-noun-0]', location='TX', pos_info='4/21', tree_codes='')\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sents = ['myocardial infarction']  # this has to be 1 sentence\n",
    "concepts,error = mm.extract_concepts(sents)\n",
    "for concept in concepts:\n",
    "    print(concept)\n",
    "    print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n",
      "True\n",
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# check char encoding, is ascii??\n",
    "    \n",
    "for ct_string in ct_final['subject_name'][:5]:\n",
    "#     print(type(ct_string))\n",
    "    print(ct_string.isascii())\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "diseases = list(set(list(ct_final['subject_name'])))\n",
    "interventions = list(filter(None, ct_final['object_name'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['pancreatic neoplasms', 'prodromal']"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diseases[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "non_ascii = \"[^\\x00-\\x7F]\"\n",
    "pattern = re.compile(r\"[^\\x00-\\x7F]\")\n",
    "\n",
    "def custom_make_translation(text):\n",
    "    non_ascii_text = re.sub(pattern, ' ', text)\n",
    "    return non_ascii_text\n",
    "\n",
    "\n",
    "text = \"split-virion, non-adjuvanted H1N1 vaccine of 15 μg\"\n",
    "translated = custom_make_translation(text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'split-virion, non-adjuvanted H1N1 vaccine of 15  g'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "translated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "diseases_translated = dict((i, custom_make_translation(i)) for i in diseases)\n",
    "interventions_translated = dict((i, custom_make_translation(i)) for i in interventions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Early Start Denver Model': 'Early Start Denver Model',\n",
       " \"'as-usual' intervention\": \"'as-usual' intervention\",\n",
       " 'Assessment of human milk nutrient composition': 'Assessment of human milk nutrient composition',\n",
       " 'KP415 oral capsule': 'KP415 oral capsule',\n",
       " 'Euglycaemic Clamp': 'Euglycaemic Clamp',\n",
       " 'Hyperinsulinaemic/Hypoglycaemic Clamp': 'Hyperinsulinaemic/Hypoglycaemic Clamp',\n",
       " 'VX-659/TEZ/IVA': 'VX-659/TEZ/IVA',\n",
       " 'TEZ/IVA': 'TEZ/IVA',\n",
       " 'IVA': 'IVA',\n",
       " 'Placebo': 'Placebo',\n",
       " 'MT-6548': 'MT-6548',\n",
       " 'Onasemnogene Abeparvovec-xioi': 'Onasemnogene Abeparvovec-xioi',\n",
       " 'Rimegepant': 'Rimegepant',\n",
       " '3D printed model': '3D printed model',\n",
       " 'Verbal Counselling': 'Verbal Counselling',\n",
       " 'Asymmetric STN-DBS': 'Asymmetric STN-DBS',\n",
       " 'Methacetin Breath Test': 'Methacetin Breath Test',\n",
       " 'Emricasan': 'Emricasan',\n",
       " 'Placebo oral capsule': 'Placebo oral capsule',\n",
       " 'High fiber sugar free biscuit': 'High fiber sugar free biscuit',\n",
       " 'CKD-501, Ketoconazole': 'CKD-501, Ketoconazole',\n",
       " 'St. Johns Wort': 'St. Johns Wort',\n",
       " 'IMPRESS intervention': 'IMPRESS intervention',\n",
       " 'EMLA': 'EMLA',\n",
       " 'Injection of Subacromial Bursa': 'Injection of Subacromial Bursa',\n",
       " 'metoprolol': 'metoprolol',\n",
       " 'Nutramigen Infant Formula': 'Nutramigen Infant Formula',\n",
       " 'AK3280': 'AK3280',\n",
       " 'Glycemic response of F-1000': 'Glycemic response of F-1000',\n",
       " 'Glycemic response to sumac sorghum flour': 'Glycemic response to sumac sorghum flour',\n",
       " 'Glycemic response to MMR sorghum flour': 'Glycemic response to MMR sorghum flour',\n",
       " 'Glycemic response to wheat flour': 'Glycemic response to wheat flour',\n",
       " 'Glycemic response to sucrose': 'Glycemic response to sucrose',\n",
       " 'Stainless Steel Crowns': 'Stainless Steel Crowns',\n",
       " 'eye patch': 'eye patch',\n",
       " 'non-eye patch': 'non-eye patch',\n",
       " 'Venues-P Valve': 'Venues-P Valve',\n",
       " 'niacin': 'niacin',\n",
       " 'placebo': 'placebo',\n",
       " 'AccuVein AV 300': 'AccuVein AV 300',\n",
       " 'MC2-01 Cream': 'MC2-01 Cream',\n",
       " 'CAL/BDP combination': 'CAL/BDP combination',\n",
       " 'GSP 301 NS': 'GSP 301 NS',\n",
       " 'GSP 301 Placebo NS': 'GSP 301 Placebo NS',\n",
       " 'Pembrolizumab': 'Pembrolizumab',\n",
       " 'Epacadostat': 'Epacadostat',\n",
       " 'Racecadotril plus ORS': 'Racecadotril plus ORS',\n",
       " 'ORS': 'ORS',\n",
       " 'Gallium Ga 68-labeled PSMA-11': 'Gallium Ga 68-labeled PSMA-11',\n",
       " 'Magnetic Resonance Imaging': 'Magnetic Resonance Imaging',\n",
       " 'Positron Emission Tomography (PET)': 'Positron Emission Tomography (PET)',\n",
       " 'Bladder instillation WITH triamcinolone acetonide': 'Bladder instillation WITH triamcinolone acetonide',\n",
       " 'Bladder instillation WITHOUT triamcinolone acetonide': 'Bladder instillation WITHOUT triamcinolone acetonide',\n",
       " 'Peer Support Dyad Intervention': 'Peer Support Dyad Intervention',\n",
       " 'Escitalopram 10mg': 'Escitalopram 10mg',\n",
       " 'Referral to Psychiatry': 'Referral to Psychiatry',\n",
       " 'Survey only': 'Survey only',\n",
       " 'ESTEEM conneCT': 'ESTEEM conneCT',\n",
       " 'Lasmiditan': 'Lasmiditan',\n",
       " 'Moxifloxacin': 'Moxifloxacin',\n",
       " 'Coffee': 'Coffee',\n",
       " 'Decaffeinated coffee': 'Decaffeinated coffee',\n",
       " 'Sugar Pill': 'Sugar Pill',\n",
       " 'Atomoxetine': 'Atomoxetine',\n",
       " 'Polihexanide': 'Polihexanide',\n",
       " 'sublingual microcirculatory evaluation': 'sublingual microcirculatory evaluation',\n",
       " 'AZD8931': 'AZD8931',\n",
       " 'Conventional technique': 'Conventional technique',\n",
       " 'Berci technique': 'Berci technique',\n",
       " 'hMG-HP/r-FSH': 'hMG-HP/r-FSH',\n",
       " 'hMG-HP': 'hMG-HP',\n",
       " 'curcumin': 'curcumin',\n",
       " 'L-Carnitine': 'L-Carnitine',\n",
       " 'ADHD Toolkit': 'ADHD Toolkit',\n",
       " 'Waiting List Control group': 'Waiting List Control group',\n",
       " 'survey': 'survey',\n",
       " 'Ustekinumab': 'Ustekinumab',\n",
       " 'VA106483': 'VA106483',\n",
       " 'Orange soda': 'Orange soda',\n",
       " 'Ocular light or darkness exposure': 'Ocular light or darkness exposure',\n",
       " 'Plumpy Nut': 'Plumpy Nut',\n",
       " 'Khichuri and Halwa': 'Khichuri and Halwa',\n",
       " 'NEEL test': 'NEEL test',\n",
       " 'ELAL test': 'ELAL test',\n",
       " 'Interview': 'Interview',\n",
       " 'cerebral fMRI': 'cerebral fMRI',\n",
       " 'droxidopa': 'droxidopa',\n",
       " 'sirolimus': 'sirolimus',\n",
       " 'Take-home Dosing Procedures': 'Take-home Dosing Procedures',\n",
       " 'Bilateral surgical implantation of DBS system to Habeluna': 'Bilateral surgical implantation of DBS system to Habeluna',\n",
       " 'Follow-up Period': 'Follow-up Period',\n",
       " 'Vitamin D': 'Vitamin D',\n",
       " 'Pulsed short wave': 'Pulsed short wave',\n",
       " 'telephone counseling 3 sessions': 'telephone counseling 3 sessions',\n",
       " 'telephone counseling 1 session': 'telephone counseling 1 session',\n",
       " 'written materials only': 'written materials only',\n",
       " 'semi-recumbent': 'semi-recumbent',\n",
       " 'supine': 'supine',\n",
       " 'Total Knee Arthroplasty with Custom Cutting Blocks': 'Total Knee Arthroplasty with Custom Cutting Blocks',\n",
       " 'Total Knee Arthroplasty via Computer Assisted Surgery': 'Total Knee Arthroplasty via Computer Assisted Surgery',\n",
       " 'Paricalcitol': 'Paricalcitol',\n",
       " 'Metronome Normal': 'Metronome Normal',\n",
       " 'Metronome Slow': 'Metronome Slow',\n",
       " 'Metronome Fast': 'Metronome Fast',\n",
       " 'Music Normal': 'Music Normal',\n",
       " 'Music Slow': 'Music Slow',\n",
       " 'Music Fast': 'Music Fast',\n",
       " 'No Auditory cueing': 'No Auditory cueing',\n",
       " 'Sirolimus': 'Sirolimus',\n",
       " 'Resveratrol': 'Resveratrol',\n",
       " 'MRgFUS treatment': 'MRgFUS treatment',\n",
       " 'Pegcetacoplan': 'Pegcetacoplan',\n",
       " 'Deep Brain Stimulation': 'Deep Brain Stimulation',\n",
       " 'Stereotatic Radiotherapy': 'Stereotatic Radiotherapy',\n",
       " 'Cytoreductive surgery combined with hyperthermic intraperitoneal chemotherapy procedure': 'Cytoreductive surgery combined with hyperthermic intraperitoneal chemotherapy procedure',\n",
       " 'Cytoreductive surgery alone': 'Cytoreductive surgery alone',\n",
       " 'avapritinib': 'avapritinib',\n",
       " 'regorafenib': 'regorafenib',\n",
       " 'Insulin Lispro': 'Insulin Lispro',\n",
       " 'LY900014': 'LY900014',\n",
       " 'Verum Cefaly® Abortive Program device': 'Verum Cefaly  Abortive Program device',\n",
       " 'Sham Cefaly® Abortive Program device': 'Sham Cefaly  Abortive Program device',\n",
       " 'Protein stomach': 'Protein stomach',\n",
       " 'Protein distal small intestine': 'Protein distal small intestine',\n",
       " 'CHO stomach': 'CHO stomach',\n",
       " 'CHO distal small intestine': 'CHO distal small intestine',\n",
       " 'Fat stomach': 'Fat stomach',\n",
       " 'Fat distal small intestine': 'Fat distal small intestine',\n",
       " 'Opti-Free : RevitaLens / Clear Care : RevitaLens': 'Opti-Free : RevitaLens / Clear Care : RevitaLens',\n",
       " 'Opti-Free : RevitaLens / RevitaLens : Clear Care': 'Opti-Free : RevitaLens / RevitaLens : Clear Care',\n",
       " 'Clear Care : RevitaLens/ Opti-Free : RevitaLens': 'Clear Care : RevitaLens/ Opti-Free : RevitaLens',\n",
       " 'Clear Care : RevitaLens / RevitaLens : Opti-Free': 'Clear Care : RevitaLens / RevitaLens : Opti-Free',\n",
       " 'RevitaLens : Clear Care/ Opti-Free : RevitaLens': 'RevitaLens : Clear Care/ Opti-Free : RevitaLens',\n",
       " 'RevitaLens : Clear Care / RevitaLens : Opti-Free': 'RevitaLens : Clear Care / RevitaLens : Opti-Free',\n",
       " 'RevitaLens : Opti-Free / Clear Care : RevitaLens': 'RevitaLens : Opti-Free / Clear Care : RevitaLens',\n",
       " 'RevitaLens : Opit-Free / RevitaLens : Clear Care': 'RevitaLens : Opit-Free / RevitaLens : Clear Care',\n",
       " 'Recombinant Human Plasma Gelsolin': 'Recombinant Human Plasma Gelsolin',\n",
       " 'Normal Saline Placebo': 'Normal Saline Placebo',\n",
       " 'Test Eye Drops': 'Test Eye Drops',\n",
       " 'Control Eye Drops': 'Control Eye Drops',\n",
       " 'KVD001 Injection': 'KVD001 Injection',\n",
       " 'Sham Procedure': 'Sham Procedure',\n",
       " 'SAPIEN 3 THV with the Commander delivery system': 'SAPIEN 3 THV with the Commander delivery system',\n",
       " 'McGrath videolaryngoscopy': 'McGrath videolaryngoscopy',\n",
       " 'Direct laryngoscopy': 'Direct laryngoscopy',\n",
       " 'Losartan potassium': 'Losartan potassium',\n",
       " 'Placebo losartan capsule': 'Placebo losartan capsule',\n",
       " 'FOL-005': 'FOL-005',\n",
       " 'FF/UMEC/VI': 'FF/UMEC/VI',\n",
       " 'Inhaled Corticosteroid': 'Inhaled Corticosteroid',\n",
       " 'LAMA': 'LAMA',\n",
       " 'LABA': 'LABA',\n",
       " 'COPD rescue medications': 'COPD rescue medications',\n",
       " 'Flortaucipir F18': 'Flortaucipir F18',\n",
       " 'Brain PET Scan': 'Brain PET Scan',\n",
       " 'Vibratory Anesthetic Device (VAD)': 'Vibratory Anesthetic Device (VAD)',\n",
       " 'Guardian RT Telemetered Glucose Monitoring System': 'Guardian RT Telemetered Glucose Monitoring System',\n",
       " 'Botox': 'Botox',\n",
       " 'Intensification of Care': 'Intensification of Care',\n",
       " 'Usual Care': 'Usual Care',\n",
       " 'Trans- resveratrol': 'Trans- resveratrol',\n",
       " 'Placebo (silica)': 'Placebo (silica)',\n",
       " 'Traditional Chinese Medicine': 'Traditional Chinese Medicine',\n",
       " 'exemestane': 'exemestane',\n",
       " 'RBT': 'RBT',\n",
       " 'Case-Management': 'Case-Management',\n",
       " 'Carotid Artery Stenting with Proximal Protection': 'Carotid Artery Stenting with Proximal Protection',\n",
       " 'Carotid Artery Stenting without Protection': 'Carotid Artery Stenting without Protection',\n",
       " 'Hawthorn': 'Hawthorn',\n",
       " 'Placebo capsule': 'Placebo capsule',\n",
       " 'Bipolar Sealer (Aquamantys)': 'Bipolar Sealer (Aquamantys)',\n",
       " 'Standard of Care': 'Standard of Care',\n",
       " 'Ipilimumab': 'Ipilimumab',\n",
       " 'electronic intervention during pregnancy': 'electronic intervention during pregnancy',\n",
       " 'Control': 'Control',\n",
       " 'electronic intervention during pregnancy and postpartum': 'electronic intervention during pregnancy and postpartum',\n",
       " 'Cognitive Behavioural Training': 'Cognitive Behavioural Training',\n",
       " 'Integrative Kinesiology Intervention': 'Integrative Kinesiology Intervention',\n",
       " 'G-CSF': 'G-CSF',\n",
       " 'Ifosfamide': 'Ifosfamide',\n",
       " 'Etoposide': 'Etoposide',\n",
       " 'Dexamethasone': 'Dexamethasone',\n",
       " 'Mesna': 'Mesna',\n",
       " 'ART-THERAPIE': 'ART-THERAPIE',\n",
       " 'EE20/DRSP (BAY86-5300)': 'EE20/DRSP (BAY86-5300)',\n",
       " 'EE20/DRSP/L-5-MTHF (BAY98-7071)': 'EE20/DRSP/L-5-MTHF (BAY98-7071)',\n",
       " 'EE20/DRSP/L-5-MTHF (BAY98-7071) + L-5-MTHF (BAY86-7660)': 'EE20/DRSP/L-5-MTHF (BAY98-7071) + L-5-MTHF (BAY86-7660)',\n",
       " 'Resolute Integrity (Zotarolimus-eluting stent)': 'Resolute Integrity (Zotarolimus-eluting stent)',\n",
       " 'Promus Element (Everolimus-eluting stent)': 'Promus Element (Everolimus-eluting stent)',\n",
       " 'Single Group Assignment': 'Single Group Assignment',\n",
       " 'No Pre-Op Testing for Cataract Surgeries': 'No Pre-Op Testing for Cataract Surgeries',\n",
       " 'hMG + GnRH antagonist': 'hMG + GnRH antagonist',\n",
       " 'APD515': 'APD515',\n",
       " 'Neuropattern™': 'Neuropattern ',\n",
       " 'Control arm': 'Control arm',\n",
       " 'Intervention arm': 'Intervention arm',\n",
       " 'Midodrine': 'Midodrine',\n",
       " 'Cohera TissuGlu Surgical Adhesive': 'Cohera TissuGlu Surgical Adhesive',\n",
       " 'Control- Standard of care': 'Control- Standard of care',\n",
       " 'Development of a diagnosis algorithm': 'Development of a diagnosis algorithm',\n",
       " 'Airway and/or pulmonary Vessels Transplantation': 'Airway and/or pulmonary Vessels Transplantation',\n",
       " 'Cognitive and behavioral Therapy': 'Cognitive and behavioral Therapy',\n",
       " 'Maximal Hyperemia with Adenosine Followed by Nicorandil': 'Maximal Hyperemia with Adenosine Followed by Nicorandil',\n",
       " 'Maximal Hyperemia with Nicorandil Followed by Adenosine': 'Maximal Hyperemia with Nicorandil Followed by Adenosine',\n",
       " 'Theravac': 'Theravac',\n",
       " 'Capecitabine (Xeloda, Roche), Oxaliplatin (Sanofi-Aventis), Docetaxel (Sanofi-Aventis)': 'Capecitabine (Xeloda, Roche), Oxaliplatin (Sanofi-Aventis), Docetaxel (Sanofi-Aventis)',\n",
       " 'AMG 386 + Paclitaxel': 'AMG 386 + Paclitaxel',\n",
       " 'AMG 386': 'AMG 386',\n",
       " 'Ultrasonic ablation device': 'Ultrasonic ablation device',\n",
       " 'Pioglitazone': 'Pioglitazone',\n",
       " 'Peripheral quantitative computed tomography (pQCT)': 'Peripheral quantitative computed tomography (pQCT)',\n",
       " 'Dual X-ray absorptiometry (DXA)': 'Dual X-ray absorptiometry (DXA)',\n",
       " 'VIMOVO (AstraZeneca)': 'VIMOVO (AstraZeneca)',\n",
       " 'VIMOVO (Patheon)': 'VIMOVO (Patheon)',\n",
       " 'Marketed enteric-coated naproxen formulation': 'Marketed enteric-coated naproxen formulation',\n",
       " 'Pulse': 'Pulse',\n",
       " 'Metformin Extended Release Oral Tablet': 'Metformin Extended Release Oral Tablet',\n",
       " 'Metformin/Gliclazide Fixed Combination': 'Metformin/Gliclazide Fixed Combination',\n",
       " 'Metformin': 'Metformin',\n",
       " 'Gliclazide': 'Gliclazide',\n",
       " 'Ascorbate': 'Ascorbate',\n",
       " 'Gemcitabine': 'Gemcitabine',\n",
       " 'Intrabone injection': 'Intrabone injection',\n",
       " 'Intra-bone cord blood hematopoietic stem cell transplantation': 'Intra-bone cord blood hematopoietic stem cell transplantation',\n",
       " 'Reminder Letter': 'Reminder Letter',\n",
       " 'Reminder Call': 'Reminder Call',\n",
       " 'Counselor Call': 'Counselor Call',\n",
       " 'continuous saphenous nerve block': 'continuous saphenous nerve block',\n",
       " 'Saline boluses in nerve catheter': 'Saline boluses in nerve catheter',\n",
       " 'Motivational Interviewing': 'Motivational Interviewing',\n",
       " 'HA wheat cereals': 'HA wheat cereals',\n",
       " 'BCT197': 'BCT197',\n",
       " 'Prednisone placebo': 'Prednisone placebo',\n",
       " 'BCT197 placebo': 'BCT197 placebo',\n",
       " 'Knee abduction moment-reducing footwear': 'Knee abduction moment-reducing footwear',\n",
       " 'Biventricular pacing': 'Biventricular pacing',\n",
       " 'No Pacing': 'No Pacing',\n",
       " 'Provent®': 'Provent ',\n",
       " 'Placebo-Provent®': 'Placebo-Provent ',\n",
       " 'CPAP': 'CPAP',\n",
       " 'Advagraf': 'Advagraf',\n",
       " 'Prograf': 'Prograf',\n",
       " 'AZD2820': 'AZD2820',\n",
       " 'Endoscopic ultrasonography': 'Endoscopic ultrasonography',\n",
       " 'everolimus': 'everolimus',\n",
       " 'erlotinib hydrochloride': 'erlotinib hydrochloride',\n",
       " 'biopsy': 'biopsy',\n",
       " 'laboratory biomarker analysis': 'laboratory biomarker analysis',\n",
       " 'immunohistochemistry staining method': 'immunohistochemistry staining method',\n",
       " 'polyacrylamide gel electrophoresis': 'polyacrylamide gel electrophoresis',\n",
       " 'pharmacological study': 'pharmacological study',\n",
       " 'external beam radiation therapy': 'external beam radiation therapy',\n",
       " 'microarray analysis': 'microarray analysis',\n",
       " 'Fluticasone propionate/salmeterol combination': 'Fluticasone propionate/salmeterol combination',\n",
       " 'Inhaled corticosteroids': 'Inhaled corticosteroids',\n",
       " 'Rosiglitazone + Metformin': 'Rosiglitazone + Metformin',\n",
       " 'Sitagliptin + Metformin': 'Sitagliptin + Metformin',\n",
       " 'Naratriptan': 'Naratriptan',\n",
       " 'Lapatinib': 'Lapatinib',\n",
       " 'Salmeterol and Fluticasone': 'Salmeterol and Fluticasone',\n",
       " 'Oatmeal': 'Oatmeal',\n",
       " 'Memantine Hydrochloride MR Prototype Capsule Formulation B': 'Memantine Hydrochloride MR Prototype Capsule Formulation B',\n",
       " 'Memantine Hydrochloride MR Prototype Capsule Formulation A': 'Memantine Hydrochloride MR Prototype Capsule Formulation A',\n",
       " 'Memantine Hydrochloride MR Prototype Capsule Formulation C': 'Memantine Hydrochloride MR Prototype Capsule Formulation C',\n",
       " 'Memantine Hydrochloride MR Prototype Capsule Formulation D': 'Memantine Hydrochloride MR Prototype Capsule Formulation D',\n",
       " 'Memantine Hydrochloride MR Prototype Capsule Formulation E': 'Memantine Hydrochloride MR Prototype Capsule Formulation E',\n",
       " 'Waterpipe smoking': 'Waterpipe smoking',\n",
       " 'This study is non-interventional.': 'This study is non-interventional.',\n",
       " 'Capecitabine': 'Capecitabine',\n",
       " 'Lu-PRRT': 'Lu-PRRT',\n",
       " 'SS-LAR': 'SS-LAR',\n",
       " 'Sensor activation (on DryMax Extra Soft)': 'Sensor activation (on DryMax Extra Soft)',\n",
       " 'ATI-50002 topical solution': 'ATI-50002 topical solution',\n",
       " 'IV Acetaminophen': 'IV Acetaminophen',\n",
       " 'PO Acetaminophen': 'PO Acetaminophen',\n",
       " 'PO Placebo': 'PO Placebo',\n",
       " 'IV Solution Placebo': 'IV Solution Placebo',\n",
       " 'tissue plasminogen activator (tPA) and human recombinant deoxyribonuclease (DNase)': 'tissue plasminogen activator (tPA) and human recombinant deoxyribonuclease (DNase)',\n",
       " 'Medical Thoracoscopy': 'Medical Thoracoscopy',\n",
       " 'PF-06763809': 'PF-06763809',\n",
       " 'Vehicle': 'Vehicle',\n",
       " 'Calcipotriene/calcipotriol': 'Calcipotriene/calcipotriol',\n",
       " 'Betamethasone': 'Betamethasone',\n",
       " 'Orlistat 120': 'Orlistat 120',\n",
       " 'Orlistat 60': 'Orlistat 60',\n",
       " 'BOTOX': 'BOTOX',\n",
       " 'Investigational Medicinal Product (MR1)': 'Investigational Medicinal Product (MR1)',\n",
       " 'Investigational Medicinal Product (MR2)': 'Investigational Medicinal Product (MR2)',\n",
       " 'Investigational Medicinal Product (MR3)': 'Investigational Medicinal Product (MR3)',\n",
       " 'Investigational Medicinal Product (MR4)': 'Investigational Medicinal Product (MR4)',\n",
       " 'Investigational Medicinal Product (MR5)': 'Investigational Medicinal Product (MR5)',\n",
       " 'Investigational Medicinal Product (IR)': 'Investigational Medicinal Product (IR)',\n",
       " 'no intervention': 'no intervention',\n",
       " 'GSK2485852 70 mg': 'GSK2485852 70 mg',\n",
       " 'GSK2485852 420 mg': 'GSK2485852 420 mg',\n",
       " 'GSK2485852 630 mg': 'GSK2485852 630 mg',\n",
       " 'GSK2485852 70 mg + Ritonavir 100mg': 'GSK2485852 70 mg + Ritonavir 100mg',\n",
       " 'GSK2485852 210 mg + Ritonavir 100mg': 'GSK2485852 210 mg + Ritonavir 100mg',\n",
       " 'GSK2485852 210 mg +Ritonavir 100mg': 'GSK2485852 210 mg +Ritonavir 100mg',\n",
       " 'GSK1349572': 'GSK1349572',\n",
       " 'Percutaneous coronary intervention': 'Percutaneous coronary intervention',\n",
       " 'GDC-0980': 'GDC-0980',\n",
       " 'bevacizumab': 'bevacizumab',\n",
       " 'capecitabine': 'capecitabine',\n",
       " 'mFOLFOX6': 'mFOLFOX6',\n",
       " 'Simvastatin,Zoledronic Acid,Bortezomib,Bendamustine,Methylprednisolone.': 'Simvastatin,Zoledronic Acid,Bortezomib,Bendamustine,Methylprednisolone.',\n",
       " 'Preimplantation Genetic Diagnosis': 'Preimplantation Genetic Diagnosis',\n",
       " 'DebDox TACE': 'DebDox TACE',\n",
       " 'Nutrition Counseling': 'Nutrition Counseling',\n",
       " 'Resistance Training and Nutrition Counseling': 'Resistance Training and Nutrition Counseling',\n",
       " 'ST101': 'ST101',\n",
       " 'high complex carbohydrates diet given in a diet group': 'high complex carbohydrates diet given in a diet group',\n",
       " 'S-1 for 6 months': 'S-1 for 6 months',\n",
       " 'S-1 for 1 year': 'S-1 for 1 year',\n",
       " 'Magnesium Sulfate': 'Magnesium Sulfate',\n",
       " 'safoof e muhazzil': 'safoof e muhazzil',\n",
       " 'compressed tablet of safoof e muhazzil': 'compressed tablet of safoof e muhazzil',\n",
       " 'Atorvastatin': 'Atorvastatin',\n",
       " 'Implantation of Holter device': 'Implantation of Holter device',\n",
       " 'Percutaneous PFO closure': 'Percutaneous PFO closure',\n",
       " 'narrow diameter heavy dental tool': 'narrow diameter heavy dental tool',\n",
       " 'large diameter light weight tool': 'large diameter light weight tool',\n",
       " 'Artery first procedure': 'Artery first procedure',\n",
       " 'No artery first procedure': 'No artery first procedure',\n",
       " 'Tigecycline': 'Tigecycline',\n",
       " 'saline nasal irrigation': 'saline nasal irrigation',\n",
       " 'Compression stockings': 'Compression stockings',\n",
       " 'compression stockings (Sigvaris )': 'compression stockings (Sigvaris )',\n",
       " 'Osteopathic Manipulative Treatment (OMT)': 'Osteopathic Manipulative Treatment (OMT)',\n",
       " 'Light touch': 'Light touch',\n",
       " 'Enriched breakfast': 'Enriched breakfast',\n",
       " 'Control breakfast': 'Control breakfast',\n",
       " 'Sequence 1': 'Sequence 1',\n",
       " 'Sequence 2': 'Sequence 2',\n",
       " 'Sequence 3': 'Sequence 3',\n",
       " 'Sequence 4': 'Sequence 4',\n",
       " 'Sequence 5': 'Sequence 5',\n",
       " 'Sequence 6': 'Sequence 6',\n",
       " '[14C] AZD5069': '[14C] AZD5069',\n",
       " 'cognitive tests and cerebral IRM': 'cognitive tests and cerebral IRM',\n",
       " 'Bevacizumab': 'Bevacizumab',\n",
       " 'BAY85-3934': 'BAY85-3934',\n",
       " 'Telaprevir': 'Telaprevir',\n",
       " 'pegIFN alfa-2a': 'pegIFN alfa-2a',\n",
       " 'Ribavirin': 'Ribavirin',\n",
       " 'cross-linking of an emulsion drink': 'cross-linking of an emulsion drink',\n",
       " 'carboplatin': 'carboplatin',\n",
       " 'RAD001': 'RAD001',\n",
       " 'paclitaxel': 'paclitaxel',\n",
       " 'chemoradiotherapy with Gemcitabine': 'chemoradiotherapy with Gemcitabine',\n",
       " 'Radiation: chemoradiotherapy with Gemcitabine': 'Radiation: chemoradiotherapy with Gemcitabine',\n",
       " 'Gemcitabine and Carboplatin': 'Gemcitabine and Carboplatin',\n",
       " 'P276-00 along with Gemcitabine and carboplatin': 'P276-00 along with Gemcitabine and carboplatin',\n",
       " 'Propranololhydrochlorid': 'Propranololhydrochlorid',\n",
       " 'Human GLP-1 (7-36) amide': 'Human GLP-1 (7-36) amide',\n",
       " 'Isotone NaCl': 'Isotone NaCl',\n",
       " 'Point-of-care \"finger-poke\" HbA1c determination.': 'Point-of-care \"finger-poke\" HbA1c determination.',\n",
       " 'HbA1c blood test': 'HbA1c blood test',\n",
       " 'HbA1c finger poke test': 'HbA1c finger poke test',\n",
       " 'Blood test': 'Blood test',\n",
       " 'Fresh lime': 'Fresh lime',\n",
       " 'Actovegin': 'Actovegin',\n",
       " 'Irbesartan': 'Irbesartan',\n",
       " 'Ciprofibrate': 'Ciprofibrate',\n",
       " 'Autologous bone marrow-derived endothelial progenitor cells': 'Autologous bone marrow-derived endothelial progenitor cells',\n",
       " 'Lifestyle behavior interventon grop': 'Lifestyle behavior interventon grop',\n",
       " 'Placebo Comparator: Control group': 'Placebo Comparator: Control group',\n",
       " 'Intermittent urinary catheterisation': 'Intermittent urinary catheterisation',\n",
       " 'parathyroid hormone (1-34)': 'parathyroid hormone (1-34)',\n",
       " 'Parathyroid Hormone-related Protein (1-36)': 'Parathyroid Hormone-related Protein (1-36)',\n",
       " 'PTH (1-34) and PTHrP (1-36)': 'PTH (1-34) and PTHrP (1-36)',\n",
       " 'Double Trouble in Recovery': 'Double Trouble in Recovery',\n",
       " 'Treatment as usual': 'Treatment as usual',\n",
       " 'injections': 'injections',\n",
       " 'Transcranial direct current stimulation (tDCS)': 'Transcranial direct current stimulation (tDCS)',\n",
       " 'Standard Dose Acclerated Fraction Radiotherapy': 'Standard Dose Acclerated Fraction Radiotherapy',\n",
       " 'Alemtuzumab': 'Alemtuzumab',\n",
       " 'Surgical Repair of Distal Radius Fracture': 'Surgical Repair of Distal Radius Fracture',\n",
       " 'Chinese herbs': 'Chinese herbs',\n",
       " 'acupuncture': 'acupuncture',\n",
       " 'Psychotherapy': 'Psychotherapy',\n",
       " 'Transcranial magnetic stimulation': 'Transcranial magnetic stimulation',\n",
       " 'Sagopilone (ZK 219477) + prednisone': 'Sagopilone (ZK 219477) + prednisone',\n",
       " 'Exercise Training': 'Exercise Training',\n",
       " 'Intravenous Vitamin C': 'Intravenous Vitamin C',\n",
       " 'Vitamin C': 'Vitamin C',\n",
       " 'Epoetin Alfa': 'Epoetin Alfa',\n",
       " 'SPM 929': 'SPM 929',\n",
       " 'topical vitamin B12': 'topical vitamin B12',\n",
       " 'Multi-Purpose Gentle Tape': 'Multi-Purpose Gentle Tape',\n",
       " 'Interpretation Modality: In Person': 'Interpretation Modality: In Person',\n",
       " 'Interpretation Modality: Telephonic': 'Interpretation Modality: Telephonic',\n",
       " 'ApneaLink Air': 'ApneaLink Air',\n",
       " 'mavacamten': 'mavacamten',\n",
       " 'rheumatoid arthritis self-management program': 'rheumatoid arthritis self-management program',\n",
       " 'SAPIEN 3 Ultra Delivery System with the SAPIEN 3 Ultra THV': 'SAPIEN 3 Ultra Delivery System with the SAPIEN 3 Ultra THV',\n",
       " 'Avatrombopag': 'Avatrombopag',\n",
       " 'Placebo Oral Tablet': 'Placebo Oral Tablet',\n",
       " 'stenfilcon A lens': 'stenfilcon A lens',\n",
       " 'narafilcon A lens': 'narafilcon A lens',\n",
       " 'Lemborexant 10 mg': 'Lemborexant 10 mg',\n",
       " 'Lemborexant 25 mg': 'Lemborexant 25 mg',\n",
       " 'direct laryngoscope': 'direct laryngoscope',\n",
       " 'video laryngoscope': 'video laryngoscope',\n",
       " 'tacrolimus ointment': 'tacrolimus ointment',\n",
       " 'Mometasone furoate': 'Mometasone furoate',\n",
       " 'Liposomal Doxorubicin': 'Liposomal Doxorubicin',\n",
       " 'Seliciclib': 'Seliciclib',\n",
       " 'Decitabine': 'Decitabine',\n",
       " 'NB-1008': 'NB-1008',\n",
       " 'aspart': 'aspart',\n",
       " 'Image-guided remove Port Wine Stain': 'Image-guided remove Port Wine Stain',\n",
       " 'Intervention Group': 'Intervention Group',\n",
       " 'Placebo Comparator: Control Group': 'Placebo Comparator: Control Group',\n",
       " 'Repetitive transcranial magnetic stimulation': 'Repetitive transcranial magnetic stimulation',\n",
       " 'IGEV': 'IGEV',\n",
       " 'curving introducer Günther Tulip Filter': 'curving introducer G nther Tulip Filter',\n",
       " 'straight introducer Günther Tulip Filter': 'straight introducer G nther Tulip Filter',\n",
       " 'Valproic acid': 'Valproic acid',\n",
       " 'Prontosan wound irrigation solution': 'Prontosan wound irrigation solution',\n",
       " 'Isotonic solution (saline or lactated ringer)': 'Isotonic solution (saline or lactated ringer)',\n",
       " 'Pemetrexed': 'Pemetrexed',\n",
       " 'IMVAMUNE®': 'IMVAMUNE ',\n",
       " 'Induction trichemotherapy - FOLFIRINOX regimen': 'Induction trichemotherapy - FOLFIRINOX regimen',\n",
       " 'Early tumor response evaluation by MRI volumetry': 'Early tumor response evaluation by MRI volumetry',\n",
       " 'Radiochemotherapy Cap 50': 'Radiochemotherapy Cap 50',\n",
       " 'Radiochemotherapy Cap 60': 'Radiochemotherapy Cap 60',\n",
       " 'Radical proctectomy with total mesorectal excision': 'Radical proctectomy with total mesorectal excision',\n",
       " 'No intervention': 'No intervention',\n",
       " 'blood collection': 'blood collection',\n",
       " 'Cardiox FDS': 'Cardiox FDS',\n",
       " 'Glucobay (Acarbose, BAYG5421)': 'Glucobay (Acarbose, BAYG5421)',\n",
       " 'Dietary Fiber Mixture': 'Dietary Fiber Mixture',\n",
       " 'Beclomethasone': 'Beclomethasone',\n",
       " 'Ciclesonide': 'Ciclesonide',\n",
       " 'Zomba Cash Transfer Program': 'Zomba Cash Transfer Program',\n",
       " 'exercise training': 'exercise training',\n",
       " 'metformin up to 2500mg/d': 'metformin up to 2500mg/d',\n",
       " 'paclitaxel 175mg/m² q21d': 'paclitaxel 175mg/m  q21d',\n",
       " 'Subcutaneous Abatacept': 'Subcutaneous Abatacept',\n",
       " 'Phenylephrine': 'Phenylephrine',\n",
       " 'Sodium-Nitroprusside': 'Sodium-Nitroprusside',\n",
       " 'Suction Cup': 'Suction Cup',\n",
       " 'Goldmann applanation tonometer': 'Goldmann applanation tonometer',\n",
       " '1060nm Optical coherence tomography': '1060nm Optical coherence tomography',\n",
       " 'PUR118': 'PUR118',\n",
       " 'Curcumin C3 tablet': 'Curcumin C3 tablet',\n",
       " 'Diacerein 1% Ointment': 'Diacerein 1% Ointment',\n",
       " 'Adelholzener Mineralwasser': 'Adelholzener Mineralwasser',\n",
       " 'Active O2': 'Active O2',\n",
       " 'Ischemic Preconditioning': 'Ischemic Preconditioning',\n",
       " 'Amoxycillin, Gentamicin': 'Amoxycillin, Gentamicin',\n",
       " 'Counselling': 'Counselling',\n",
       " 'Peer counseling on infant feeding': 'Peer counseling on infant feeding',\n",
       " 'Blood sampling': 'Blood sampling',\n",
       " 'Biopsy': 'Biopsy',\n",
       " 'Conventional Surgery': 'Conventional Surgery',\n",
       " 'Genetic Testing': 'Genetic Testing',\n",
       " 'Biotronik Vision Guidewire': 'Biotronik Vision Guidewire',\n",
       " 'GS-9131': 'GS-9131',\n",
       " 'BIC': 'BIC',\n",
       " 'TAF': 'TAF',\n",
       " 'ARV regimen': 'ARV regimen',\n",
       " 'DRV': 'DRV',\n",
       " 'RTV': 'RTV',\n",
       " 'Acetaminophen IV/PO': 'Acetaminophen IV/PO',\n",
       " 'Acetaminophen PO': 'Acetaminophen PO',\n",
       " 'Ketorolac': 'Ketorolac',\n",
       " 'Celebrex': 'Celebrex',\n",
       " 'Naproxen': 'Naproxen',\n",
       " 'Tramadol': 'Tramadol',\n",
       " 'Pregabalin': 'Pregabalin',\n",
       " 'Gabapentin': 'Gabapentin',\n",
       " 'Lidocaine': 'Lidocaine',\n",
       " 'Opioids': 'Opioids',\n",
       " 'Regional anesthesia': 'Regional anesthesia',\n",
       " 'diacerein 1% ointment': 'diacerein 1% ointment',\n",
       " 'Danicopan': 'Danicopan',\n",
       " 'Eculizumab': 'Eculizumab',\n",
       " 'Nasal non-invasive ventilation with RAM cannula': 'Nasal non-invasive ventilation with RAM cannula',\n",
       " 'laparoscopic salpingotomy or salpingostomy': 'laparoscopic salpingotomy or salpingostomy',\n",
       " 'Ethanol-lock': 'Ethanol-lock',\n",
       " 'Diacerein 1% ointment': 'Diacerein 1% ointment',\n",
       " 'observational': 'observational',\n",
       " 'GB002': 'GB002',\n",
       " 'Generic Dry Powder Inhaler': 'Generic Dry Powder Inhaler',\n",
       " 'Infusion of allogeneic umbilical cord blood': 'Infusion of allogeneic umbilical cord blood',\n",
       " 'Infusion of MSCs': 'Infusion of MSCs',\n",
       " 'Indomethacin': 'Indomethacin',\n",
       " 'Diclofenac': 'Diclofenac',\n",
       " 'Meloxicam': 'Meloxicam',\n",
       " 'Celecoxib': 'Celecoxib',\n",
       " 'task-practice therapy': 'task-practice therapy',\n",
       " 'Vibration': 'Vibration',\n",
       " 'Intervention arm: PP-MI': 'Intervention arm: PP-MI',\n",
       " 'Stable rowing simulator': 'Stable rowing simulator',\n",
       " 'Slides based rowing simulator': 'Slides based rowing simulator',\n",
       " 'Tilt board based rowing simulator': 'Tilt board based rowing simulator',\n",
       " 'Exposure': 'Exposure',\n",
       " 'CANscript': 'CANscript',\n",
       " 'NNC 0123-0000-0338': 'NNC 0123-0000-0338',\n",
       " 'insulin glargine': 'insulin glargine',\n",
       " 'DC-006 vaccine': 'DC-006 vaccine',\n",
       " 'p.DOM-WT1-37 DNA Vaccine and p.DOM-WT1-126 DNA Vaccine': 'p.DOM-WT1-37 DNA Vaccine and p.DOM-WT1-126 DNA Vaccine',\n",
       " 'Axitinib plus everolimus': 'Axitinib plus everolimus',\n",
       " 'SMArTVIEW': 'SMArTVIEW',\n",
       " 'Strategy Training': 'Strategy Training',\n",
       " 'Decision aid': 'Decision aid',\n",
       " 'CP-675,206': 'CP-675,206',\n",
       " 'External local radiation therapy': 'External local radiation therapy',\n",
       " 'Axitinib (AG-013736)': 'Axitinib (AG-013736)',\n",
       " 'A 4-week diet, low in sodium.': 'A 4-week diet, low in sodium.',\n",
       " 'Poly-amide': 'Poly-amide',\n",
       " 'Poly-methyl methacrylate': 'Poly-methyl methacrylate',\n",
       " 'Nutrient Intake': 'Nutrient Intake',\n",
       " 'Tiotropium': 'Tiotropium',\n",
       " 'Albuterol/salbutamol': 'Albuterol/salbutamol',\n",
       " 'Placebo to match FF/UMEC/VI': 'Placebo to match FF/UMEC/VI',\n",
       " 'Placebo to match tiotropium': 'Placebo to match tiotropium',\n",
       " 'ELLIPTA inhaler': 'ELLIPTA inhaler',\n",
       " 'HANDIHALER': 'HANDIHALER',\n",
       " 'MDI': 'MDI',\n",
       " 'Cloak Shape Device for Sham Tuina': 'Cloak Shape Device for Sham Tuina',\n",
       " 'Cloak Shape Device for Genuine Tuina': 'Cloak Shape Device for Genuine Tuina',\n",
       " 'follow up system and multimodality approach': 'follow up system and multimodality approach',\n",
       " 'NeoMatriX Collagen Dressing': 'NeoMatriX Collagen Dressing',\n",
       " 'cAd3-Marburg vaccine': 'cAd3-Marburg vaccine',\n",
       " 'Social Dancing': 'Social Dancing',\n",
       " 'Treadmill Walking': 'Treadmill Walking',\n",
       " 'senofilcon A TEST Lens': 'senofilcon A TEST Lens',\n",
       " 'senofilcon A CONTROL Lens': 'senofilcon A CONTROL Lens',\n",
       " 'Insulin LISPRO': 'Insulin LISPRO',\n",
       " 'recombinant human insulin': 'recombinant human insulin',\n",
       " 'Recovery Management Checkups': 'Recovery Management Checkups',\n",
       " 'Outcome Monitoring': 'Outcome Monitoring',\n",
       " 'TLR8 agonist VTX-2337': 'TLR8 agonist VTX-2337',\n",
       " 'cetuximab': 'cetuximab',\n",
       " 'pharmacogenomic studies': 'pharmacogenomic studies',\n",
       " 'Occipital neurolysis': 'Occipital neurolysis',\n",
       " 'Low-power microwave breast imaging system': 'Low-power microwave breast imaging system',\n",
       " 'MenACYW conjugate vaccine': 'MenACYW conjugate vaccine',\n",
       " 'Questionnaire': 'Questionnaire',\n",
       " 'Audiometry': 'Audiometry',\n",
       " 'Urine biomarkers for renal tubular function': 'Urine biomarkers for renal tubular function',\n",
       " 'Blood pressure': 'Blood pressure',\n",
       " 'Rucaparib': 'Rucaparib',\n",
       " 'QL Block': 'QL Block',\n",
       " 'TAP Block': 'TAP Block',\n",
       " 'Whole blood': 'Whole blood',\n",
       " 'Enhanced Couple HIV Testing and Counseling': 'Enhanced Couple HIV Testing and Counseling',\n",
       " 'FreeStyle LibrePro CGM': 'FreeStyle LibrePro CGM',\n",
       " 'CeFAZolin 1000 MG': 'CeFAZolin 1000 MG',\n",
       " 'Clindamycin 900 MG in 6 ML Injection': 'Clindamycin 900 MG in 6 ML Injection',\n",
       " 'NAFL-treatment': 'NAFL-treatment',\n",
       " 'Smart Aging Program': 'Smart Aging Program',\n",
       " 'Educational Materials': 'Educational Materials',\n",
       " 'AL-611': 'AL-611',\n",
       " 'Ranolazine': 'Ranolazine',\n",
       " 'Schedule A': 'Schedule A',\n",
       " 'Schedule B': 'Schedule B',\n",
       " 'Schedule C': 'Schedule C',\n",
       " 'Schedule D': 'Schedule D',\n",
       " 'Schedule E': 'Schedule E',\n",
       " 'iSONEP (sonepcizumab/LT1009)': 'iSONEP (sonepcizumab/LT1009)',\n",
       " 'Taxus Liberte Paclitaxel-Eluting Coronary Stent System': 'Taxus Liberte Paclitaxel-Eluting Coronary Stent System',\n",
       " 'Resolute Zotarolimus-Eluting Coronary Stent System': 'Resolute Zotarolimus-Eluting Coronary Stent System',\n",
       " 'Blood Draw': 'Blood Draw',\n",
       " 'ForeseeHome': 'ForeseeHome',\n",
       " 'Lung Volume Reduction Coil (LVRC) (PneumRx)': 'Lung Volume Reduction Coil (LVRC) (PneumRx)',\n",
       " 'elective neck dissection': 'elective neck dissection',\n",
       " 'Tacrolimus - Prograf® twice daily formulation': 'Tacrolimus - Prograf  twice daily formulation',\n",
       " 'Tacrolimus - Advagraf® once daily formulation': 'Tacrolimus - Advagraf  once daily formulation',\n",
       " 'Nike Pegasus': 'Nike Pegasus',\n",
       " 'Nike Free 3.0': 'Nike Free 3.0',\n",
       " 'Vibram Five Fingers Shoe': 'Vibram Five Fingers Shoe',\n",
       " 'Physical Activity': 'Physical Activity',\n",
       " 'CAMS': 'CAMS',\n",
       " 'Health & Wellness': 'Health & Wellness',\n",
       " 'propofol': 'propofol',\n",
       " 'Genetic testing': 'Genetic testing',\n",
       " 'TMS , H2-coil': 'TMS , H2-coil',\n",
       " 'Arabin Cervical Pessary': 'Arabin Cervical Pessary',\n",
       " 'rituximab': 'rituximab',\n",
       " 'cyclophosphamide': 'cyclophosphamide',\n",
       " 'doxorubicin hydrochloride': 'doxorubicin hydrochloride',\n",
       " 'prednisone': 'prednisone',\n",
       " 'vincristine sulfate': 'vincristine sulfate',\n",
       " 'SAFE VET': 'SAFE VET',\n",
       " 'Tenofovir DF': 'Tenofovir DF',\n",
       " 'Cognitive-Behavioral Therapy for Pain and Opioid Dependence': 'Cognitive-Behavioral Therapy for Pain and Opioid Dependence',\n",
       " 'Methadone Drug Counseling': 'Methadone Drug Counseling',\n",
       " 'Active Arm N-acetyl cysteine (NAC)': 'Active Arm N-acetyl cysteine (NAC)',\n",
       " 'Placebo Arm': 'Placebo Arm',\n",
       " 'appropriate volume of 0.5% ropivacaine': 'appropriate volume of 0.5% ropivacaine',\n",
       " 'Continuous interscalene block': 'Continuous interscalene block',\n",
       " 'PCA morphine': 'PCA morphine',\n",
       " 'Copeptin': 'Copeptin',\n",
       " 'Phacoemulsification cataract surgery': 'Phacoemulsification cataract surgery',\n",
       " 'Glucose-bicarbonate-Ringer Lactate (GBRL)': 'Glucose-bicarbonate-Ringer Lactate (GBRL)',\n",
       " 'Balanced Salt Solution (BSS®)': 'Balanced Salt Solution (BSS )',\n",
       " 'recombinant human cardiotrophin-1': 'recombinant human cardiotrophin-1',\n",
       " 'Usual care': 'Usual care',\n",
       " 'Lung age': 'Lung age',\n",
       " 'Contingency Management': 'Contingency Management',\n",
       " 'Lung age + Contingency Management': 'Lung age + Contingency Management',\n",
       " 'Letrozole/IUI': 'Letrozole/IUI',\n",
       " 'CC/IUI': 'CC/IUI',\n",
       " 'Open anterior inguinal hernia repair': 'Open anterior inguinal hernia repair',\n",
       " 'sleep deprivation': 'sleep deprivation',\n",
       " 'Normal sleep': 'Normal sleep',\n",
       " 'Influenza vaccination': 'Influenza vaccination',\n",
       " 'Pneumococcal vaccination': 'Pneumococcal vaccination',\n",
       " 'Hepatitis B Virus (HBV) vaccination': 'Hepatitis B Virus (HBV) vaccination',\n",
       " 'tampon D': 'tampon D',\n",
       " 'tampon M': 'tampon M',\n",
       " 'tampon T': 'tampon T',\n",
       " 'tampon V': 'tampon V',\n",
       " 'Experimental Denture Wipe': 'Experimental Denture Wipe',\n",
       " 'Tap Water': 'Tap Water',\n",
       " 'Placebo DPI': 'Placebo DPI',\n",
       " 'ELLIPTA DPI': 'ELLIPTA DPI',\n",
       " 'Ease of use questionnaires': 'Ease of use questionnaires',\n",
       " 'budesonide/formoterol': 'budesonide/formoterol',\n",
       " 'albuterol/salbutamol': 'albuterol/salbutamol',\n",
       " 'Placebo to match budesonide/formoterol': 'Placebo to match budesonide/formoterol',\n",
       " 'tiotropium': 'tiotropium',\n",
       " 'placebo to match tiotropium': 'placebo to match tiotropium',\n",
       " 'ELLIPTA': 'ELLIPTA',\n",
       " 'HandiHaler': 'HandiHaler',\n",
       " 'risankizumab': 'risankizumab',\n",
       " 'secukinumab': 'secukinumab',\n",
       " 'Lumiracoxib': 'Lumiracoxib',\n",
       " 'valsartan + amlodipine': 'valsartan + amlodipine',\n",
       " 'MMF': 'MMF',\n",
       " 'BCNU': 'BCNU',\n",
       " 'VP-16': 'VP-16',\n",
       " 'CY': 'CY',\n",
       " 'FTBI': 'FTBI',\n",
       " 'BU': 'BU',\n",
       " 'Omega-3 fatty acids': 'Omega-3 fatty acids',\n",
       " 'Home for Life mobile application': 'Home for Life mobile application',\n",
       " 'SR-1 Laser': 'SR-1 Laser',\n",
       " 'Deep Brain Stimulation (Model 7438 Therapy Controller)': 'Deep Brain Stimulation (Model 7438 Therapy Controller)',\n",
       " 'Polar M400 + Facebook Group (PM400+FG)': 'Polar M400 + Facebook Group (PM400+FG)',\n",
       " 'Facebook Only Group (FG)': 'Facebook Only Group (FG)',\n",
       " 'Bicarbonated mineral water': 'Bicarbonated mineral water',\n",
       " 'Mineral water low in mineral content (control)': 'Mineral water low in mineral content (control)',\n",
       " 'No drug will be used': 'No drug will be used',\n",
       " 'ranibizumab 2.0 mg': 'ranibizumab 2.0 mg',\n",
       " 'Cyclosporine Inhalation Solution': 'Cyclosporine Inhalation Solution',\n",
       " 'YM178': 'YM178',\n",
       " 'Large field of view arm': 'Large field of view arm',\n",
       " 'Medium field of view': 'Medium field of view',\n",
       " 'REST': 'REST',\n",
       " 'lecture': 'lecture',\n",
       " 'deferasirox': 'deferasirox',\n",
       " 'atorvastatin and clopidogrel': 'atorvastatin and clopidogrel',\n",
       " 'Clopidogrel': 'Clopidogrel',\n",
       " 'Temsirolimus + Sorafenib': 'Temsirolimus + Sorafenib',\n",
       " 'Standard care': 'Standard care',\n",
       " 'continuous positive airway pressure': 'continuous positive airway pressure',\n",
       " 'Emulsion type artificial tear': 'Emulsion type artificial tear',\n",
       " 'Sublingual immunotherapy (SLIT)': 'Sublingual immunotherapy (SLIT)',\n",
       " 'Subcutaneous immunotherapy (SCIT)': 'Subcutaneous immunotherapy (SCIT)',\n",
       " 'Coping with Breast Cancer web-based workbook': 'Coping with Breast Cancer web-based workbook',\n",
       " 'ALXN1102': 'ALXN1102',\n",
       " 'ALXN1103': 'ALXN1103',\n",
       " 'Intervention': 'Intervention',\n",
       " 'Transcutaneous Electrical Nerve Stimulator (TENS)': 'Transcutaneous Electrical Nerve Stimulator (TENS)',\n",
       " 'The Transforaminal Lumbar Interbody Fusion (TLIF) surgery': 'The Transforaminal Lumbar Interbody Fusion (TLIF) surgery',\n",
       " 'Early Rehabilitation Program In Critically Ill Chronic Obstructed Airway Disease patients': 'Early Rehabilitation Program In Critically Ill Chronic Obstructed Airway Disease patients',\n",
       " 'Meclizine Oral Tablet': 'Meclizine Oral Tablet',\n",
       " 'Obeticholic acid': 'Obeticholic acid',\n",
       " 'Placebos': 'Placebos',\n",
       " 'Normal saline': 'Normal saline',\n",
       " 'BAY1000394': 'BAY1000394',\n",
       " 'BI 853520': 'BI 853520',\n",
       " 'Flavocoxid': 'Flavocoxid',\n",
       " 'Hylan GF-20 alone': 'Hylan GF-20 alone',\n",
       " 'Triamcinolone': 'Triamcinolone',\n",
       " 'bismuth subcitrate potassium, metronidazole,tetracycline hydrochloride and omeprazole': 'bismuth subcitrate potassium, metronidazole,tetracycline hydrochloride and omeprazole',\n",
       " 'Acceptance and commitment therapy': 'Acceptance and commitment therapy',\n",
       " 'ND1.1': 'ND1.1',\n",
       " 'Placebo control': 'Placebo control',\n",
       " 'Nevirapine': 'Nevirapine',\n",
       " 'Metronidazole/Miconazole (Neo-Penotran Forte, BAY86-5276)': 'Metronidazole/Miconazole (Neo-Penotran Forte, BAY86-5276)',\n",
       " 'Chlorobenzene Liquid': 'Chlorobenzene Liquid',\n",
       " 'KLYX': 'KLYX',\n",
       " 'Glycerine': 'Glycerine',\n",
       " 'Japanese Encephalitis Virus vaccine, inactivated, adsorbed': 'Japanese Encephalitis Virus vaccine, inactivated, adsorbed',\n",
       " 'Cementless (Nexgen)': 'Cementless (Nexgen)',\n",
       " 'Cemented (Nexgen)': 'Cemented (Nexgen)',\n",
       " 'AZD5213': 'AZD5213',\n",
       " 'ExAblate 2000': 'ExAblate 2000',\n",
       " 'Laser Refractive Surgery': 'Laser Refractive Surgery',\n",
       " 'Motivational enhancement therapy': 'Motivational enhancement therapy',\n",
       " 'Marijuana education': 'Marijuana education',\n",
       " 'nicotine transdermal patch/lozenge': 'nicotine transdermal patch/lozenge',\n",
       " 'VRC-EBOMAB092-00-AB (MAb114)': 'VRC-EBOMAB092-00-AB (MAb114)',\n",
       " 'Enzalutamide Liquid': 'Enzalutamide Liquid',\n",
       " 'Enzalutamide Capsule': 'Enzalutamide Capsule',\n",
       " 'Omalizumab': 'Omalizumab',\n",
       " 'Accu-Chek® Solo micropump system': 'Accu-Chek  Solo micropump system',\n",
       " 'mylife™ OmniPod® Insulin Management System': 'mylife  OmniPod  Insulin Management System',\n",
       " 'Multiple Daily Injections (MDI) therapy': 'Multiple Daily Injections (MDI) therapy',\n",
       " 'Staccato Alprazolam': 'Staccato Alprazolam',\n",
       " 'Precedex': 'Precedex',\n",
       " 'FLACS': 'FLACS',\n",
       " 'Manual conventional surgery': 'Manual conventional surgery',\n",
       " 'CENTURION® Vision System': 'CENTURION  Vision System',\n",
       " 'LenSx®': 'LenSx ',\n",
       " 'Dextromethorphan Hydrobromide': 'Dextromethorphan Hydrobromide',\n",
       " 'Avicel PH101 (Microcrystalline Cellulose NF) for Compounding': 'Avicel PH101 (Microcrystalline Cellulose NF) for Compounding',\n",
       " 'Oxycodone': 'Oxycodone',\n",
       " 'Liraglutide Pen Injector [Saxenda]': 'Liraglutide Pen Injector [Saxenda]',\n",
       " 'Placebo Liraglutide Pen Injector': 'Placebo Liraglutide Pen Injector',\n",
       " 'Breastfeeding and weight loss support': 'Breastfeeding and weight loss support',\n",
       " 'Cook stoves, either Biolite wood stove, or liquified petroleum gas stove plus fuel': 'Cook stoves, either Biolite wood stove, or liquified petroleum gas stove plus fuel',\n",
       " 'AN-PEP': 'AN-PEP',\n",
       " 'Glypressin (terlipressin)': 'Glypressin (terlipressin)',\n",
       " 'Boceprevir, Peg-interferon alfa 2b and Ribavirin': 'Boceprevir, Peg-interferon alfa 2b and Ribavirin',\n",
       " 'Hepatocyte Matrix Implant': 'Hepatocyte Matrix Implant',\n",
       " 'clinical, endoscopical and manometric evaluation': 'clinical, endoscopical and manometric evaluation',\n",
       " 'Samatasvir tablet': 'Samatasvir tablet',\n",
       " 'Samatasvir capsule': 'Samatasvir capsule',\n",
       " 'Lumbar Microdiscectomy': 'Lumbar Microdiscectomy',\n",
       " 'Physiotherapy, Epidural injections, Education, Pain Medications, Anti-inflammatories': 'Physiotherapy, Epidural injections, Education, Pain Medications, Anti-inflammatories',\n",
       " 'Fractional flow reserve': 'Fractional flow reserve',\n",
       " 'IVUS': 'IVUS',\n",
       " 'ETV；TDF；ADV': 'ETV TDF ADV',\n",
       " 'Peginterferon Alfa-2a': 'Peginterferon Alfa-2a',\n",
       " 'Confocal Laser Endomicroscopy': 'Confocal Laser Endomicroscopy',\n",
       " 'repetitive magnetic stimulation': 'repetitive magnetic stimulation',\n",
       " 'Ixazomib': 'Ixazomib',\n",
       " 'Melphalan': 'Melphalan',\n",
       " 'Prednisone': 'Prednisone',\n",
       " 'ChronVac-C + SOC': 'ChronVac-C + SOC',\n",
       " 'SOC': 'SOC',\n",
       " 'aerobic training': 'aerobic training',\n",
       " 'Wait list': 'Wait list',\n",
       " 'Insulin glargine': 'Insulin glargine',\n",
       " 'Mindfulness-Based Stress Reduction': 'Mindfulness-Based Stress Reduction',\n",
       " 'Cognitive Behavior Therapy': 'Cognitive Behavior Therapy',\n",
       " 'DHEA': 'DHEA',\n",
       " 'Growth Hormone-Releasing Hormone': 'Growth Hormone-Releasing Hormone',\n",
       " 'Placebo 1': 'Placebo 1',\n",
       " 'Placebo 2': 'Placebo 2',\n",
       " 'Early Time-Restricted Feeding': 'Early Time-Restricted Feeding',\n",
       " 'Late Time-Restricted Feeding': 'Late Time-Restricted Feeding',\n",
       " 'AZD9291': 'AZD9291',\n",
       " 'National Cancer Institute (NCI)-4650, a messenger ribonucleic acid (mRNA)-based, Personalized Cancer Vaccine': 'National Cancer Institute (NCI)-4650, a messenger ribonucleic acid (mRNA)-based, Personalized Cancer Vaccine',\n",
       " 'Padsevonil (UCB0942)': 'Padsevonil (UCB0942)',\n",
       " 'Erythromycin': 'Erythromycin',\n",
       " 'Measuring extubation force': 'Measuring extubation force',\n",
       " 'IVUS and OCT Imaging': 'IVUS and OCT Imaging',\n",
       " 'trientine dihydrochloride': 'trientine dihydrochloride',\n",
       " 'pegylated liposomal doxorubicin': 'pegylated liposomal doxorubicin',\n",
       " 'Medical ultrasound device (ShearWave system)': 'Medical ultrasound device (ShearWave system)',\n",
       " 'Navigation station coupled with a 3D fluoroscopy': 'Navigation station coupled with a 3D fluoroscopy',\n",
       " 'Web-based Access to CPAP Data': 'Web-based Access to CPAP Data',\n",
       " 'Web-based Access to CPAP Data with Financial Incentive': 'Web-based Access to CPAP Data with Financial Incentive',\n",
       " 'Coupons for use within the next week': 'Coupons for use within the next week',\n",
       " 'Coupons for use within the next 3 months': 'Coupons for use within the next 3 months',\n",
       " 'Indocyanine Green': 'Indocyanine Green',\n",
       " 'Information': 'Information',\n",
       " 'Cryoablation': 'Cryoablation',\n",
       " 'DMUC5754A': 'DMUC5754A',\n",
       " 'Telemonitoring': 'Telemonitoring',\n",
       " 'Telemonitoring & Telemedicine': 'Telemonitoring & Telemedicine',\n",
       " 'control group': 'control group',\n",
       " 'Metformin Pill': 'Metformin Pill',\n",
       " 'Vildagliptin 50 mg': 'Vildagliptin 50 mg',\n",
       " 'VIM detection': 'VIM detection',\n",
       " 'Blood is obtained, mostly during routine venipuncture': 'Blood is obtained, mostly during routine venipuncture',\n",
       " 'Hypnosis': 'Hypnosis',\n",
       " 'continuous glucose monitoring (CGM)': 'continuous glucose monitoring (CGM)',\n",
       " 'SHP607': 'SHP607',\n",
       " 'Saline': 'Saline',\n",
       " 'Bupivacaine liposome': 'Bupivacaine liposome',\n",
       " 'SC-007': 'SC-007',\n",
       " 'SIMMAX2': 'SIMMAX2',\n",
       " 'Oxaliplatin': 'Oxaliplatin',\n",
       " 'LV5FU or Folfiri': 'LV5FU or Folfiri',\n",
       " 'WBV exercise program': 'WBV exercise program',\n",
       " 'Sham WBV exercise program': 'Sham WBV exercise program',\n",
       " 'Tailored regimen of IV PCA according to pain sensitivity': 'Tailored regimen of IV PCA according to pain sensitivity',\n",
       " 'Regimen of IV PCA without considering pain sensitivity': 'Regimen of IV PCA without considering pain sensitivity',\n",
       " 'vaginal swab -microbiome test and blood test for cytokines': 'vaginal swab -microbiome test and blood test for cytokines',\n",
       " 'Peginterferon alfa-2a': 'Peginterferon alfa-2a',\n",
       " 'Nimotuzumab': 'Nimotuzumab',\n",
       " 'Nanoparticle Albumin-Bound Paclitaxel': 'Nanoparticle Albumin-Bound Paclitaxel',\n",
       " 'Percutaneous radiofrequency catheter ablation': 'Percutaneous radiofrequency catheter ablation',\n",
       " 'Mini invasive thoracoscopic radiofrequency ablation': 'Mini invasive thoracoscopic radiofrequency ablation',\n",
       " 'ADX48621': 'ADX48621',\n",
       " 'EPIC™ Self-Expanding Nitinol Vascular Stent': 'EPIC  Self-Expanding Nitinol Vascular Stent',\n",
       " 'Occupational Therapy': 'Occupational Therapy',\n",
       " 'MLC601': 'MLC601',\n",
       " 'split-virion, non-adjuvanted H1N1 vaccine of 15 μg': 'split-virion, non-adjuvanted H1N1 vaccine of 15  g',\n",
       " 'split-virion, non-adjuvanted H1N1 vaccine of 30 μg': 'split-virion, non-adjuvanted H1N1 vaccine of 30  g',\n",
       " 'split-virion, non-adjuvanted H1N1 vaccine of 45 μg': 'split-virion, non-adjuvanted H1N1 vaccine of 45  g',\n",
       " 'Salvadora persica': 'Salvadora persica',\n",
       " 'Toothbrush': 'Toothbrush',\n",
       " 'Best supportive care': 'Best supportive care',\n",
       " 'fostamatinib': 'fostamatinib',\n",
       " 'rifampicin': 'rifampicin',\n",
       " 'GLPG0778': 'GLPG0778',\n",
       " 'SEBASTIAN Clinical Decision Support System (CDSS)': 'SEBASTIAN Clinical Decision Support System (CDSS)',\n",
       " 'biological collection': 'biological collection',\n",
       " 'Endurance training': 'Endurance training',\n",
       " 'Strength training': 'Strength training',\n",
       " 'Combined training': 'Combined training',\n",
       " 'Yoga Research Tool': 'Yoga Research Tool',\n",
       " 'Sitagliptin': 'Sitagliptin',\n",
       " 'Pembrolizumab plus amurubicin': 'Pembrolizumab plus amurubicin',\n",
       " 'Pure Encapsulations Branched Chain Amino Acid powder': 'Pure Encapsulations Branched Chain Amino Acid powder',\n",
       " 'control group (no BCAAs provided)': 'control group (no BCAAs provided)',\n",
       " 'WO 4260 Cosmetic Product for Topical Use': 'WO 4260 Cosmetic Product for Topical Use',\n",
       " 'uterus transplantation': 'uterus transplantation',\n",
       " 'Prasugrel': 'Prasugrel',\n",
       " 'Tirofiban': 'Tirofiban',\n",
       " 'Survey': 'Survey',\n",
       " 'Chemotherapy': 'Chemotherapy',\n",
       " 'TBI-1401(HF10)': 'TBI-1401(HF10)',\n",
       " 'Nab-paclitaxel': 'Nab-paclitaxel',\n",
       " 'TS-1': 'TS-1',\n",
       " 'BMS-986165': 'BMS-986165',\n",
       " 'V114': 'V114',\n",
       " 'Prevnar 13™': 'Prevnar 13 ',\n",
       " 'PNEUMOVAX™23': 'PNEUMOVAX 23',\n",
       " 'Language ENhancement Assessment/intervention system': 'Language ENhancement Assessment/intervention system',\n",
       " 'Shen Men acupuncture': 'Shen Men acupuncture',\n",
       " 'Epidural injection': 'Epidural injection',\n",
       " 'Sofosbuvir': 'Sofosbuvir',\n",
       " 'Daclatasvir': 'Daclatasvir',\n",
       " 'Pegylated Interferon alfa-2a': 'Pegylated Interferon alfa-2a',\n",
       " 'Concordant': 'Concordant',\n",
       " 'Cyclophosphamide': 'Cyclophosphamide',\n",
       " 'Elotuzumab': 'Elotuzumab',\n",
       " 'Laboratory Biomarker Analysis': 'Laboratory Biomarker Analysis',\n",
       " 'Lenalidomide': 'Lenalidomide',\n",
       " 'Pharmacological Study': 'Pharmacological Study',\n",
       " 'Herbst Appliance': 'Herbst Appliance',\n",
       " 'Distal Jet Appliance': 'Distal Jet Appliance',\n",
       " 'DEXA': 'DEXA',\n",
       " 'pQCT': 'pQCT',\n",
       " 'Lifestyle Questionnaire': 'Lifestyle Questionnaire',\n",
       " 'Aerobic Fitness': 'Aerobic Fitness',\n",
       " 'Muscular Strength': 'Muscular Strength',\n",
       " 'Blood draw and finger prick test': 'Blood draw and finger prick test',\n",
       " 'Muscular Power': 'Muscular Power',\n",
       " 'Endotracheal Intubation': 'Endotracheal Intubation',\n",
       " 'anodal tDCS': 'anodal tDCS',\n",
       " 'cathodal tDCS': 'cathodal tDCS',\n",
       " 'sham tDCS': 'sham tDCS',\n",
       " 'HealthLoop mobile application': 'HealthLoop mobile application',\n",
       " 'Blood sample': 'Blood sample',\n",
       " 'Survey therapists': 'Survey therapists',\n",
       " 'Survey patients': 'Survey patients',\n",
       " 'ACTIVE peripheral nerve stimulation with a percutaneously inserted lead and wearable stimulator (SPR Therapeutics, Cleveland, Ohio)': 'ACTIVE peripheral nerve stimulation with a percutaneously inserted lead and wearable stimulator (SPR Therapeutics, Cleveland, Ohio)',\n",
       " 'SHAM peripheral nerve stimulation with a percutaneously inserted lead and wearable SHAM stimulator (SPR Therapeutics, Cleveland, Ohio)': 'SHAM peripheral nerve stimulation with a percutaneously inserted lead and wearable SHAM stimulator (SPR Therapeutics, Cleveland, Ohio)',\n",
       " 'ETBX-071; adenoviral PSA vaccine': 'ETBX-071; adenoviral PSA vaccine',\n",
       " 'ETBX-061; adenoviral MUC1 vaccine': 'ETBX-061; adenoviral MUC1 vaccine',\n",
       " 'ETBX-051; adenoviral brachyury vaccine': 'ETBX-051; adenoviral brachyury vaccine',\n",
       " 'retinoid 9cUAB30': 'retinoid 9cUAB30',\n",
       " 'cancer prevention': 'cancer prevention',\n",
       " 'single cell haplotyping': 'single cell haplotyping',\n",
       " 'Botulinum toxin type A (Meditoxin®)': 'Botulinum toxin type A (Meditoxin )',\n",
       " 'DCE MRI': 'DCE MRI',\n",
       " 'Infusion of donor derived ex-vivo selected NK cells to patients after transplant': 'Infusion of donor derived ex-vivo selected NK cells to patients after transplant',\n",
       " 'Haematology / Blood chemistry sampling': 'Haematology / Blood chemistry sampling',\n",
       " 'Chemotherapy following concurrent Chemoradiotherapy': 'Chemotherapy following concurrent Chemoradiotherapy',\n",
       " 'Oral appliance': 'Oral appliance',\n",
       " 'taxotere': 'taxotere',\n",
       " 'Zirconium (Zirconium Genesis II)': 'Zirconium (Zirconium Genesis II)',\n",
       " 'Cobalt-Chrome (Cobalt Chrome Genesis II)': 'Cobalt-Chrome (Cobalt Chrome Genesis II)',\n",
       " 'Retigabine IR': 'Retigabine IR',\n",
       " \"Equine F(ab')2 antivenom\": \"Equine F(ab')2 antivenom\",\n",
       " 'Intensive care support plus placebo': 'Intensive care support plus placebo',\n",
       " 'WRSs2': 'WRSs2',\n",
       " 'WRSs3': 'WRSs3',\n",
       " 'Teaching life style changes': 'Teaching life style changes',\n",
       " 'Active participation': 'Active participation',\n",
       " 'Teaching': 'Teaching',\n",
       " '75% insulin lispro protamine suspension and 25 % insulin lispro injection': '75% insulin lispro protamine suspension and 25 % insulin lispro injection',\n",
       " 'Ariva Silver Wintergreen': 'Ariva Silver Wintergreen',\n",
       " 'Placebo Wintergreen': 'Placebo Wintergreen',\n",
       " 'ETR/telaprevir': 'ETR/telaprevir',\n",
       " 'TMC278/telaprevir': 'TMC278/telaprevir',\n",
       " 'telaprevir/ETR': 'telaprevir/ETR',\n",
       " 'telaprevir/TMC278': 'telaprevir/TMC278',\n",
       " 'Panobinostat, Cisplatin, Pemetrexed': 'Panobinostat, Cisplatin, Pemetrexed',\n",
       " 'RNA analysis': 'RNA analysis',\n",
       " 'gene expression analysis': 'gene expression analysis',\n",
       " 'nucleic acid sequencing': 'nucleic acid sequencing',\n",
       " 'diagnostic laboratory biomarker analysis': 'diagnostic laboratory biomarker analysis',\n",
       " 'French adaptation of NINDS-Canadian Stroke Network battery': 'French adaptation of NINDS-Canadian Stroke Network battery',\n",
       " 'carfilzomib': 'carfilzomib',\n",
       " 'Psycho educational and behavioural group sessions': 'Psycho educational and behavioural group sessions',\n",
       " 'Walking session': 'Walking session',\n",
       " 'Irinotecan': 'Irinotecan',\n",
       " 'Chemoembolization': 'Chemoembolization',\n",
       " 'Drug-Eluting Beads': 'Drug-Eluting Beads',\n",
       " 'arginine': 'arginine',\n",
       " 'autologous CD133pos stem cell application': 'autologous CD133pos stem cell application',\n",
       " 'VATS lobectomy using modified equipments': 'VATS lobectomy using modified equipments',\n",
       " 'VATS lobectomy using less staples': 'VATS lobectomy using less staples',\n",
       " 'Standard VATS lobectomy': 'Standard VATS lobectomy',\n",
       " 'Open lobectomy': 'Open lobectomy',\n",
       " 'Mentored medication reconciliation quality improvement': 'Mentored medication reconciliation quality improvement',\n",
       " 'Observation of compliance with the World Health Organisation Surgical Safety Checklist': 'Observation of compliance with the World Health Organisation Surgical Safety Checklist',\n",
       " 'Mirikizumab': 'Mirikizumab',\n",
       " 'Physician Receives QoL results': 'Physician Receives QoL results',\n",
       " 'Physician Does not receive the results': 'Physician Does not receive the results',\n",
       " 'Telecare system': 'Telecare system',\n",
       " 'blood samples': 'blood samples',\n",
       " 'Line extension of the Phonak Audéo B product family': 'Line extension of the Phonak Aud o B product family',\n",
       " 'vascular duplex ultrasound imaging': 'vascular duplex ultrasound imaging',\n",
       " 'TAK-788': 'TAK-788',\n",
       " 'Nicotine Replacement Therapy': 'Nicotine Replacement Therapy',\n",
       " 'Referral to AHEC/Quitline': 'Referral to AHEC/Quitline',\n",
       " 'Spesolimab': 'Spesolimab',\n",
       " 'Gefapixant 45 mg': 'Gefapixant 45 mg',\n",
       " 'Pimavanserin': 'Pimavanserin',\n",
       " 'Bupivacaine Group': 'Bupivacaine Group',\n",
       " 'BI 1467335': 'BI 1467335',\n",
       " 'BI 1467335 (C-14) intravenous solution': 'BI 1467335 (C-14) intravenous solution',\n",
       " 'NATO WELP combination': 'NATO WELP combination',\n",
       " 'NATO FIS combination': 'NATO FIS combination',\n",
       " 'RAVEN WELP combination': 'RAVEN WELP combination',\n",
       " 'RAVEN FIS combination': 'RAVEN FIS combination',\n",
       " 'Tamibarotene': 'Tamibarotene',\n",
       " 'High-intensity interval training': 'High-intensity interval training',\n",
       " 'Resistance training': 'Resistance training',\n",
       " 'Pelvic floor exercises with biofeedback': 'Pelvic floor exercises with biofeedback',\n",
       " 'Pelvic floor muscle exercises': 'Pelvic floor muscle exercises',\n",
       " 'Stenting with ELLA Biodegradable stent': 'Stenting with ELLA Biodegradable stent',\n",
       " 'Standard Dilations': 'Standard Dilations',\n",
       " 'Tobramycin': 'Tobramycin',\n",
       " 'Complex sentence treatment protocol': 'Complex sentence treatment protocol',\n",
       " 'Dutasteride': 'Dutasteride',\n",
       " 'SBT- PSV': 'SBT- PSV',\n",
       " 'SBT - NAVA': 'SBT - NAVA',\n",
       " 'Tesetaxel': 'Tesetaxel',\n",
       " 'Fluticasone furoate nasal spray': 'Fluticasone furoate nasal spray',\n",
       " 'Pentoxifylline': 'Pentoxifylline',\n",
       " 'Insulin clamp': 'Insulin clamp',\n",
       " 'peginterferon alfa 2a [Pegasys]': 'peginterferon alfa 2a [Pegasys]',\n",
       " 'peginterferon alfa-2a [Pegasys]': 'peginterferon alfa-2a [Pegasys]',\n",
       " 'Cediranib': 'Cediranib',\n",
       " 'VMS diary booklet': 'VMS diary booklet',\n",
       " 'BIIB017': 'BIIB017',\n",
       " 'UDCA': 'UDCA',\n",
       " 'oxygenated oil': 'oxygenated oil',\n",
       " 'medical device. Spray containing essentially oxygenated oil': 'medical device. Spray containing essentially oxygenated oil',\n",
       " 'Outdoor use of plain distance glasses with counselling': 'Outdoor use of plain distance glasses with counselling',\n",
       " 'Mesalamine': 'Mesalamine',\n",
       " 'Laser Preipheral Iridotomy and phacoemulsification': 'Laser Preipheral Iridotomy and phacoemulsification',\n",
       " 'Sildenafil Citrate': 'Sildenafil Citrate',\n",
       " 'rapamycin-eluting Stent with permanent polymer': 'rapamycin-eluting Stent with permanent polymer',\n",
       " 'rapamycin-eluting stent with biodegradable polymer': 'rapamycin-eluting stent with biodegradable polymer',\n",
       " 'polymer-free, rapamycin-eluting stent': 'polymer-free, rapamycin-eluting stent',\n",
       " 'ER OROS paliperidone and Olanzapine': 'ER OROS paliperidone and Olanzapine',\n",
       " 'Provera': 'Provera',\n",
       " 'Combination Birth Control pills': 'Combination Birth Control pills',\n",
       " '64 Channel VCT': '64 Channel VCT',\n",
       " 'brief (behavioral therapy)': 'brief (behavioral therapy)',\n",
       " 'Drop out re-engagement motivational intervention': 'Drop out re-engagement motivational intervention',\n",
       " 'standard care follow up': 'standard care follow up',\n",
       " 'Mindfulness Telehealth Intervention': 'Mindfulness Telehealth Intervention',\n",
       " 'Psychoeducation Telehealth Intervention': 'Psychoeducation Telehealth Intervention',\n",
       " 'Granulocyte-Macrophage Colony-Stimulating Factor (GM-CSF)': 'Granulocyte-Macrophage Colony-Stimulating Factor (GM-CSF)',\n",
       " 'Treatment as usual (TAU)': 'Treatment as usual (TAU)',\n",
       " 'CBT for CBT': 'CBT for CBT',\n",
       " 'Standard CBT': 'Standard CBT',\n",
       " 'CBT+CM/adherence': 'CBT+CM/adherence',\n",
       " 'CM/abstinence': 'CM/abstinence',\n",
       " 'CM/abstinence+CBT': 'CM/abstinence+CBT',\n",
       " 'Deferiprone (L1)': 'Deferiprone (L1)',\n",
       " 'Desferrioxamine': 'Desferrioxamine',\n",
       " 'Helical tomotherapy IMRT': 'Helical tomotherapy IMRT',\n",
       " 'MBT and BreathID': 'MBT and BreathID',\n",
       " 'Erlotinib and bevacizumab': 'Erlotinib and bevacizumab',\n",
       " 'Autologous Bone Marrow Mononuclear Cells (ABMMC) Transplantation': 'Autologous Bone Marrow Mononuclear Cells (ABMMC) Transplantation',\n",
       " 'EEG': 'EEG',\n",
       " 'Azacitidine': 'Azacitidine',\n",
       " 'XL820': 'XL820',\n",
       " 'Extended-Release Carbamazepine': 'Extended-Release Carbamazepine',\n",
       " 'Cognitive Therapy': 'Cognitive Therapy',\n",
       " 'Treatment as Usual': 'Treatment as Usual',\n",
       " 'Thumb orthoses': 'Thumb orthoses',\n",
       " 'brief intervention (cognitive-behavioral therapy)': 'brief intervention (cognitive-behavioral therapy)',\n",
       " 'Psychoeducation': 'Psychoeducation',\n",
       " 'fMRI': 'fMRI',\n",
       " 'Telcyta': 'Telcyta',\n",
       " 'a strategy based on PCT guided AB therapy': 'a strategy based on PCT guided AB therapy',\n",
       " 'Immunosuppressive Therapy': 'Immunosuppressive Therapy',\n",
       " 'Survey Administration': 'Survey Administration',\n",
       " 'Daylight Intervention': 'Daylight Intervention',\n",
       " 'Microwave treatment': 'Microwave treatment',\n",
       " 'CHIKV VLP/unadjuvanted': 'CHIKV VLP/unadjuvanted',\n",
       " ...}"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interventions_translated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# test translation function on test list\n",
    "\n",
    "# diseases_translated = dict((i, custom_make_translation(i)) for i in diseases[:20])\n",
    "# interventions_translated = dict((i, custom_make_translation(i)) for i in interventions[:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def run_metamap(input_term, source_restriction):\n",
    "#     concepts_dict = dict()\n",
    "#     try:\n",
    "#         concepts,error = mm.extract_concepts(sentences = [input_term], restrict_to_sources = source_restriction)\n",
    "#         concepts_dict[input_term] = concepts\n",
    "#     except:\n",
    "#         concepts_dict[input_term] = None \n",
    "#     return(concepts_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# original MetaMap function\n",
    "\n",
    "# def run_metamap_orig(input_term):\n",
    "#     concepts_dict = dict()\n",
    "#     try:\n",
    "#         concepts,error = mm.extract_concepts(sentences = [input_term])\n",
    "#         concepts_dict[input_term] = concepts\n",
    "#     except:\n",
    "#         concepts_dict[input_term] = None \n",
    "#     return(concepts_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_metamap(input_term, source_restriction):\n",
    "    concepts_dict = dict()\n",
    "    if all(x is None for x in source_restriction):\n",
    "        try:\n",
    "            concepts,error = mm.extract_concepts([input_term],\n",
    "                                                 word_sense_disambiguation = True,\n",
    "                                                 prune = 10,\n",
    "                                                 composite_phrase = 1)\n",
    "            concepts_dict[input_term] = concepts\n",
    "        except:\n",
    "            concepts_dict[input_term] = None \n",
    "    else:\n",
    "        try:\n",
    "            concepts,error = mm.extract_concepts([input_term],\n",
    "                                                 word_sense_disambiguation = True,\n",
    "                                                 restrict_to_sources=source_restriction,\n",
    "                                                 prune = 10,\n",
    "                                                 composite_phrase = 1)\n",
    "            concepts_dict[input_term] = concepts\n",
    "        except:\n",
    "            concepts_dict[input_term] = None \n",
    "    return(concepts_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "avail_num_workers = multiprocessing.cpu_count() - 1  \n",
    "\n",
    "# pool = multiprocessing.Pool(num_workers)\n",
    "# for task in tasks:\n",
    "#     pool.apply_async(func, args = (task,))\n",
    "\n",
    "# pool.close()\n",
    "# pool.join()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "# uncomment to run MetaMap on full dataset from Clinical Trials\n",
    "# number_of_processes = 4\n",
    "# metamapped_diseases = multiprocessing.Pool(multiprocessing.cpu_count() - 1).starmap(run_metamap,\n",
    "#                                                                         zip(list(diseases_translated.values()), [['SNOMEDCT_US', 'SNOMEDCT_VET', 'ICD10CM', 'ICD10CM', 'ICD10PCS', 'ICD9CM', 'ICD9CM']]*len(list(diseases_translated.values()))))\n",
    "# metamapped_interventions = multiprocessing.Pool(multiprocessing.cpu_count() - 1).starmap(run_metamap, \n",
    "#                                                                              zip(list(interventions_translated.values()),\n",
    "#                                                                                  [[None]]*len(list(interventions_translated.values()))))\n",
    "                                                                             \n",
    "# Stop MetaMap servers\n",
    "os.system(metamap_base_dir + metamap_pos_server_dir + ' stop') # Part of speech tagger\n",
    "os.system(metamap_base_dir + metamap_wsd_server_dir + ' stop') # Word sense disambiguation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'metamapped_diseases' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m----------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-34-f0e14aebd617>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mmetamapped_diseases\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmetamapped_diseases\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mtest_dict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmetamapped_diseases\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mtest_dict\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'metamapped_diseases' is not defined"
     ]
    }
   ],
   "source": [
    "metamapped_diseases[0]\n",
    "type(metamapped_diseases[0])\n",
    "test_dict = metamapped_diseases[:2]\n",
    "test_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # CAREFUL: UNCOMMENTING WILL OVERWRITE METAMAPPED DISEASES AND INTERVENTIONS FILES, WHICH TAKE FOREVER TO GENERATE!\n",
    "\n",
    "# # write mapped diseases and interventions to pickle file\n",
    "# with open('metamapped_diseases.txt', 'wb') as file:\n",
    "#     pickle.dump(metamapped_diseases, file)\n",
    "# with open('metamapped_interventions.txt', 'wb') as file:\n",
    "#     pickle.dump(metamapped_interventions, file)\n",
    "\n",
    "# # write mapped diseases and interventions to human readable file\n",
    "# with open('readable_metamapped_diseases.txt', 'w') as file:\n",
    "#     file.write(json.dumps(metamapped_diseases, indent=4))\n",
    "# with open('readable_metamapped_interventions.txt', 'w') as file:\n",
    "#     file.write(json.dumps(metamapped_interventions, indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('metamapped_diseases.txt', 'rb') as f:\n",
    "    metamapped_diseases = pickle.load(f)\n",
    "    \n",
    "with open('metamapped_interventions.txt', 'rb') as f:\n",
    "    metamapped_interventions = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "Can not infer schema for type: <class 'str'>",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m----------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-128-222abf3c81d1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;31m# dataframe = spark.createDataFrame(data, col)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 22\u001b[0;31m \u001b[0mdiseases_df_mapped\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspark\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreateDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     23\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/opt/apache-spark/libexec/python/pyspark/sql/session.py\u001b[0m in \u001b[0;36mcreateDataFrame\u001b[0;34m(self, data, schema, samplingRatio, verifySchema)\u001b[0m\n\u001b[1;32m    673\u001b[0m             return super(SparkSession, self).createDataFrame(\n\u001b[1;32m    674\u001b[0m                 data, schema, samplingRatio, verifySchema)\n\u001b[0;32m--> 675\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_create_dataframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msamplingRatio\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverifySchema\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    676\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    677\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_create_dataframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msamplingRatio\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverifySchema\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/opt/apache-spark/libexec/python/pyspark/sql/session.py\u001b[0m in \u001b[0;36m_create_dataframe\u001b[0;34m(self, data, schema, samplingRatio, verifySchema)\u001b[0m\n\u001b[1;32m    698\u001b[0m             \u001b[0mrdd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_createFromRDD\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprepare\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msamplingRatio\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    699\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 700\u001b[0;31m             \u001b[0mrdd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_createFromLocal\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprepare\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    701\u001b[0m         \u001b[0mjrdd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jvm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSerDeUtil\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtoJavaArray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_to_java_object_rdd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    702\u001b[0m         \u001b[0mjdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jsparkSession\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapplySchemaToPythonRDD\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrdd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mschema\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjson\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/opt/apache-spark/libexec/python/pyspark/sql/session.py\u001b[0m in \u001b[0;36m_createFromLocal\u001b[0;34m(self, data, schema)\u001b[0m\n\u001b[1;32m    510\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    511\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mschema\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mschema\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 512\u001b[0;31m             \u001b[0mstruct\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_inferSchemaFromList\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnames\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mschema\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    513\u001b[0m             \u001b[0mconverter\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_create_converter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstruct\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    514\u001b[0m             \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconverter\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/opt/apache-spark/libexec/python/pyspark/sql/session.py\u001b[0m in \u001b[0;36m_inferSchemaFromList\u001b[0;34m(self, data, names)\u001b[0m\n\u001b[1;32m    437\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    438\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"can not infer schema from empty dataset\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 439\u001b[0;31m         \u001b[0mschema\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mreduce\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_merge_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0m_infer_schema\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnames\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    440\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0m_has_nulltype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mschema\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    441\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Some of types cannot be determined after inferring\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/opt/apache-spark/libexec/python/pyspark/sql/session.py\u001b[0m in \u001b[0;36m<genexpr>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    437\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    438\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"can not infer schema from empty dataset\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 439\u001b[0;31m         \u001b[0mschema\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mreduce\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_merge_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0m_infer_schema\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnames\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    440\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0m_has_nulltype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mschema\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    441\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Some of types cannot be determined after inferring\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/opt/apache-spark/libexec/python/pyspark/sql/types.py\u001b[0m in \u001b[0;36m_infer_schema\u001b[0;34m(row, names)\u001b[0m\n\u001b[1;32m   1065\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1066\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1067\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Can not infer schema for type: %s\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1068\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1069\u001b[0m     \u001b[0mfields\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: Can not infer schema for type: <class 'str'>"
     ]
    }
   ],
   "source": [
    "# print(metamapped_diseases[:2])\n",
    "# print(\"\\n\")\n",
    "# print({kv[0]:kv[1] for i, kv in enumerate(diseases_translated.items()) if i <= 4})\n",
    "\n",
    "# import org.apache.spark.sql.types.{StringType, StructField, StructType}\n",
    "\n",
    "schema = StructType(\n",
    "    Array(\n",
    "        StructField(\"original_ct_value\", StringType,true),\n",
    "        StructField(\"metamap_input\", StringType,true),\n",
    "        StructField(\"metamap_output_preferred_name\", StringType,true),\n",
    "        StructField(\"cui\", StringType,true),\n",
    "        StructField(\"score\", StringType,true),\n",
    "        StructField(\"semtypes\", StringType,true),\n",
    "        \n",
    "))\n",
    "\n",
    "col = ['original_ct_value', 'metamap_input', 'metamap_output_preferred_name', 'cui', 'score', 'semtypes' ]\n",
    "\n",
    "# dataframe = spark.createDataFrame(data, col)\n",
    "\n",
    "# diseases_df_mapped = spark.createDataFrame(col)\n",
    "\n",
    "\n",
    "for disease in metamapped_diseases[:10]:\n",
    "        for key in disease:\n",
    "            for concept in disease[key]:\n",
    "                concept = concept._asdict()\n",
    "    #             concept.get('preferred_name')\n",
    "                row = [key]\n",
    "                original_disease = (list(diseases_translated.keys())[list(diseases_translated.values()).index(key)])\n",
    "                row.append(original_disease)\n",
    "                row.extend([concept.get(k) for k in ['preferred_name', 'cui', 'score', 'semtypes']])\n",
    "                print(row)\n",
    "                diseases_df_mapped = diseases_df_mapped.union(row)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diseases_translated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['infertile high-risk patients with polycystic ovaries', 'infertile high-risk patients with polycystic ovaries', 'Polycystic Ovary Syndrome', 'C0032460', '22.36', '[dsyn]']\n",
      "['anaesthetic induction', 'anaesthetic induction', 'Anesthetics', 'C0002932', '19.25', '[phsu]']\n",
      "['riss stage ii plasma cell myeloma', 'riss stage ii plasma cell myeloma', 'Multiple Myeloma', 'C0026764', '16.42', '[neop]']\n",
      "['metabolic abnormality', 'metabolic abnormality', 'Abnormal', 'C0205161', '3.55', '[fndg]']\n",
      "['chalazion unspecified eye, unspecified eyelid', 'chalazion unspecified eye, unspecified eyelid', 'Eyelid structure', 'C0015426', '16.30', '[bpoc]']\n",
      "['refractory transformed marginal zone lymphoma to diffuse large b-cell lymphoma', 'refractory transformed marginal zone lymphoma to diffuse large b-cell lymphoma', 'Diffuse Large B-Cell Lymphoma', 'C0079744', '22.58', '[neop]']\n",
      "['internet based wellness program', 'internet based wellness program', 'Alkalies', 'C0002055', '7.13', '[inch]']\n",
      "['infant colics', 'infant colics', 'Infantile Colic', 'C0266836', '9.97', '[sosy]']\n",
      "['abgar score of the neonate', 'abgar score of the neonate', 'Infant, Newborn', 'C0021289', '12.89', '[aggp]']\n",
      "['pain in a joint, multiple sites', 'pain in a joint, multiple sites', 'Arthralgia', 'C0003862', '16.27', '[sosy]']\n",
      "['sperm capacitation', 'sperm capacitation', 'sperm cell', 'C0037868', '12.94', '[cell]']\n",
      "['optical enhancement endoscopy', 'optical enhancement endoscopy', 'Vision', 'C0042789', '16.05', '[orgf]']\n",
      "['hypertension in the obstetric context', 'hypertension in the obstetric context', 'Non-proteinuric hypertension of pregnancy', 'C1443005', '3.85', '[dsyn]']\n",
      "['broad solid tumor', 'broad solid tumor', 'Neoplasms', 'C0027651', '3.61', '[neop]']\n",
      "['very high risk neuroblastoma', 'very high risk neuroblastoma', 'Neuroblastoma', 'C0027819', '25.68', '[neop]']\n",
      "['locally advanced non small cell lung cancer (nsclc)*', 'locally advanced non small cell lung cancer (nsclc)*', None, None, None, None]\n",
      "['gastritis acute', 'gastritis acute', 'Gastritis', 'C0017152', '13.14', '[dsyn]']\n",
      "['renal anemia, chronic', 'renal anemia, chronic', 'Anemia', 'C0002871', '9.94', '[dsyn]']\n",
      "['invasive epithelial ovarian cancer', 'invasive epithelial ovarian cancer', 'Malignant neoplasm of ovary', 'C1140680', '3.65', '[neop]']\n",
      "['heart valves', 'heart valves', 'Heart Valves', 'C0018826', '10.06', '[bpoc]']\n",
      "['bilirubin-induced neurological dysfunction (bind)', 'bilirubin-induced neurological dysfunction (bind)', None, None, None, None]\n",
      "['ultilization of out-patient colonoscopy service', 'ultilization of out-patient colonoscopy service', 'colonoscopy', 'C0009378', '22.35', '[diap]']\n",
      "['primary enuresis', 'primary enuresis', 'primary enuresis', 'C0232847', '3.75', '[sosy]']\n",
      "['qtc evaluation', 'qtc evaluation', 'Evaluation procedure', 'C1261322', '3.63', '[hlca]']\n",
      "['pregnancy-related pelvic girdle pain', 'pregnancy-related pelvic girdle pain', 'Pelvic Girdle Pain', 'C0240682', '19.45', '[fndg]']\n",
      "['localized provoked vulvodynia', 'localized provoked vulvodynia', 'Vulvodynia', 'C0406670', '16.30', '[dsyn]']\n",
      "['pneumocystis carinii pneumonia', 'pneumocystis carinii pneumonia', 'Pneumocystis jiroveci pneumonia', 'C1535939', '13.27', '[dsyn]']\n",
      "['acute necrotizing pancreatitis', 'acute necrotizing pancreatitis', 'Pancreatitis, Acute Necrotizing', 'C0267941', '13.27', '[dsyn]']\n",
      "['progression from impaired fasting glucose to diabetes mellitus', 'progression from impaired fasting glucose to diabetes mellitus', 'Diabetes Mellitus', 'C0011849', '14.64', '[dsyn]']\n",
      "['stage iia prostate cancer ajcc v8', 'stage iia prostate cancer ajcc v8', 'Malignant neoplasm of prostate', 'C0376358', '3.47', '[neop]']\n",
      "['b-cell neoplasm', 'b-cell neoplasm', 'B-Cell Lymphomas', 'C0079731', '19.58', '[neop]']\n",
      "['carbohydrate', 'carbohydrate', 'Carbohydrates', 'C0007004', '3.68', '[orch]']\n",
      "['education, professional', 'education, professional', 'Educational Status', 'C0013658', '9.99', '[fndg]']\n",
      "['craniosynostoses', 'craniosynostoses', 'Craniosynostosis', 'C0010278', '19.45', '[dsyn]']\n",
      "['change; circulatory', 'change; circulatory', 'Cardiovascular system', 'C0007226', '5.18', '[bdsy]']\n",
      "['coronary occlusion', 'coronary occlusion', 'Coronary Occlusion', 'C0151814', '16.37', '[dsyn]']\n",
      "['multiple family groups by parent peers', 'multiple family groups by parent peers', 'parent', 'C0030551', '19.34', '[famg]']\n",
      "['dexamethasone effect on mirna profile in healthy humans', 'dexamethasone effect on mirna profile in healthy humans', 'Homo sapiens', 'C0086418', '28.91', '[humn]']\n",
      "['chronic coronary occlusion', 'chronic coronary occlusion', 'Coronary Occlusion', 'C0151814', '16.30', '[dsyn]']\n",
      "['patients with major burn injury in formosa funcost explosion', 'patients with major burn injury in formosa funcost explosion', 'Burn injury', 'C0006434', '13.26', '[inpo]']\n",
      "['circulating tumor cells (ctcs)', 'circulating tumor cells (ctcs)', 'Tumor cells, uncertain whether benign or malignant', 'C0431085', '3.68', '[cell]']\n",
      "['locally advanced or metastatic cancer', 'locally advanced or metastatic cancer', 'Advanced phase', 'C0205179', '5.18', '[qlco]']\n",
      "['nerve compression syndrome', 'nerve compression syndrome', 'Syndrome', 'C0039082', '13.08', '[dsyn]']\n",
      "['beta-cell function', 'beta-cell function', 'Structure of beta Cell of islet', 'C0030281', '13.14', '[cell]']\n",
      "['cardiovascular risks', 'cardiovascular risks', 'Cardiovascular system', 'C0007226', '3.63', '[bdsy]']\n",
      "['active non-segmental vitiligo', 'active non-segmental vitiligo', 'Segmental vitiligo', 'C1274648', '3.65', '[dsyn]']\n",
      "['lung lesion(s) requiring evaluation', 'lung lesion(s) requiring evaluation', 'Lesion of lung', 'C0577916', '3.75', '[fndg]']\n",
      "['treadmill', 'treadmill', 'Treadmill, device', 'C0184069', '3.68', '[mnob]']\n",
      "['stage ib non-small cell lung carcinoma', 'stage ib non-small cell lung carcinoma', 'Non-Small Cell Lung Carcinoma', 'C0007131', '25.81', '[neop]']\n",
      "['ebv+ sarcomas', 'ebv+ sarcomas', 'Epstein-Barr Virus', 'C0014644', '19.25', '[virs]']\n",
      "['cardiac transplantion', 'cardiac transplantion', 'Heart', 'C0018787', '6.63', '[bpoc]']\n",
      "['mcrpc or advanced/metastatic solid tumors', 'mcrpc or advanced/metastatic solid tumors', 'Advanced phase', 'C0205179', '5.18', '[qlco]']\n",
      "['complete uterine evacuation after use of study drugs', 'complete uterine evacuation after use of study drugs', 'Drug Evaluation', 'C0013175', '9.74', '[resa]']\n",
      "['skin perfusion', 'skin perfusion', 'Perfusion (procedure)', 'C0031001', '6.79', '[topp]']\n",
      "['substance related disorders', 'substance related disorders', 'Disease', 'C0012634', '9.92', '[dsyn]']\n",
      "['chemosensitive hodgkin lymphoma', 'chemosensitive hodgkin lymphoma', 'Hodgkin Disease', 'C0019829', '16.30', '[neop]']\n",
      "['cushing; syndrome or disease, glucocorticoid-induced', 'cushing; syndrome or disease, glucocorticoid-induced', 'Glucocorticoids', 'C0017710', '19.27', '[horm,orch]']\n",
      "['jaw spasm', 'jaw spasm', 'Trismus', 'C0041105', '19.52', '[dsyn]']\n",
      "['surgical site infection in obese women undergoing c-section', 'surgical site infection in obese women undergoing c-section', 'Obesity', 'C0028754', '28.66', '[dsyn]']\n",
      "['muscle loss in elderly', 'muscle loss in elderly', 'Elderly (population group)', 'C0001792', '12.89', '[popg]']\n",
      "['patello femoral syndrome', 'patello femoral syndrome', 'Patellofemoral stress syndrome', 'C2315536', '3.81', '[dsyn]']\n",
      "['hungry-satiation situation', 'hungry-satiation situation', 'Hunger', 'C0020175', '12.93', '[sosy]']\n",
      "['outdoor activity', 'outdoor activity', 'Active', 'C0205177', '3.63', '[qlco]']\n",
      "['diaphragmatic weakness', 'diaphragmatic weakness', 'Diaphragmatic paresis', 'C0521532', '3.75', '[fndg]']\n",
      "['non diabetic subjects with normoglycemia', 'non diabetic subjects with normoglycemia', 'Blood glucose normal', 'C0580545', '3.38', '[fndg]']\n",
      "['esophagoscopy techniques', 'esophagoscopy techniques', 'Esophagoscopy', 'C0014873', '22.40', '[diap]']\n",
      "['genu valgum or varum', 'genu valgum or varum', 'Knee joint valgus deformity', 'C0576093', '10.06', '[anab]']\n",
      "['single embryo transfer', 'single embryo transfer', 'Unmarried person', 'C0037179', '22.54', '[popg]']\n",
      "['anal high-grade squamous intraepithelial lesions', 'anal high-grade squamous intraepithelial lesions', 'High-Grade Squamous Intraepithelial Lesions', 'C0333875', '22.71', '[neop]']\n",
      "['immune tolerance', 'immune tolerance', 'Immune Tolerance', 'C0020963', '10.06', '[patf]']\n",
      "['partial aniridia', 'partial aniridia', 'Aniridia', 'C0003076', '13.09', '[cgab]']\n",
      "['dwarfism', 'dwarfism', 'Dwarfism', 'C0013336', '13.14', '[cgab]']\n",
      "['spine osteoarthritis', 'spine osteoarthritis', 'Osteoarthritis, Spine', 'C2350242', '19.52', '[dsyn]']\n",
      "['b-cell lymphomas', 'b-cell lymphomas', 'B-Cell Lymphomas', 'C0079731', '19.58', '[neop]']\n",
      "['high grade b-cell lymphoma w/myc & bcl2 or bcl6 rearrangements', 'high grade b-cell lymphoma w/myc & bcl2 or bcl6 rearrangements', 'High grade B-cell lymphoma', 'C0456863', '3.51', '[neop]']\n",
      "['hereditary colorectal cancer', 'hereditary colorectal cancer', 'Malignant neoplasm of large intestine', 'C0346629', '3.68', '[neop]']\n",
      "['stage iii ovarian yolk sac tumor ajcc v6 and v7', 'stage iii ovarian yolk sac tumor ajcc v6 and v7', 'Endodermal sinus tumor of ovary', 'C0346188', '3.49', '[neop]']\n",
      "['polypoidal choroidal vasculopathy without active polyp', 'polypoidal choroidal vasculopathy without active polyp', 'Choroid', 'C0008520', '12.95', '[bpoc]']\n",
      "['alveolar bone grafting', 'alveolar bone grafting', 'Alveolar Bone Grafting', 'C0407708', '16.42', '[topp]']\n",
      "['pediatric procedures', 'pediatric procedures', 'Pediatrics', 'C0030755', '9.78', '[bmod]']\n",
      "['immuno-monitoring', 'immuno-monitoring', 'Monitoring - action', 'C1283169', '3.63', '[hlca]']\n",
      "['non gist sarcomas', 'non gist sarcomas', 'Gastrointestinal Stromal Tumors', 'C0238198', '16.08', '[neop]']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['enzyme disorder', 'enzyme disorder', 'Enzymopathy', 'C0520572', '3.75', '[dsyn]']\n",
      "['atypical burkitt/burkitt-like lymphoma', 'atypical burkitt/burkitt-like lymphoma', 'Burkitt-like lymphoma', 'C1368771', '3.88', '[neop]']\n",
      "['adult fibrosarcoma', 'adult fibrosarcoma', 'Fibrosarcoma', 'C0016057', '19.40', '[neop]']\n",
      "['burkitts lymphoma', 'burkitts lymphoma', 'Burkitt Lymphoma', 'C0006413', '22.68', '[neop]']\n",
      "['closed-loop glucose control', 'closed-loop glucose control', 'Glucose', 'C0017725', '16.22', '[bacs,orch,phsu]']\n",
      "['skeletal muscle physiology', 'skeletal muscle physiology', 'Skeletal muscle structure', 'C0242692', '12.96', '[tisu]']\n",
      "['neoplasms, neck', 'neoplasms, neck', 'Neck', 'C0027530', '8.34', '[blor]']\n",
      "['mind-body-medicine', 'mind-body-medicine', 'Body structure', 'C1268086', '3.61', '[anst]']\n",
      "['evaluating impact of pacv on vaccine hesitancy among parents of newborns', 'evaluating impact of pacv on vaccine hesitancy among parents of newborns', 'parent', 'C0030551', '19.19', '[famg]']\n",
      "['nummular dermatitis', 'nummular dermatitis', 'Nummular eczema', 'C0085656', '3.75', '[dsyn]']\n",
      "['recurrent ependymoma', 'recurrent ependymoma', 'Ependymoma', 'C0014474', '22.56', '[neop]']\n",
      "['biopsy-proven iga nephropathy', 'biopsy-proven iga nephropathy', 'IGA Glomerulonephritis', 'C0017661', '22.57', '[dsyn]']\n",
      "['subdural hematoma, non-acute', 'subdural hematoma, non-acute', 'Hematoma, Subdural', 'C0018946', '22.68', '[patf]']\n",
      "['adult acute myeloid leukemia with t(8;21); (q22; q22.1); runx1-runx1t1', 'adult acute myeloid leukemia with t(8;21); (q22; q22.1); runx1-runx1t1', 'Leukemia, Myelocytic, Acute', 'C0023467', '16.22', '[neop]']\n",
      "['pearson syndrome', 'pearson syndrome', \"Pearson's marrow-pancreas syndrome\", 'C0342784', '13.22', '[dsyn]']\n",
      "['competitive behavior', 'competitive behavior', 'Behavior', 'C0004927', '6.79', '[bhvr]']\n",
      "['sleep-related leg cramps', 'sleep-related leg cramps', 'Muscle Cramp', 'C0026821', '17.80', '[sosy]']\n"
     ]
    }
   ],
   "source": [
    "# write the mapping results to text files for evaluation \n",
    "\n",
    "with open(\"metamapped_diseases_qual_control.csv\", \"w\") as file:\n",
    "    csv.excel.delimiter=','\n",
    "    writer = csv.writer(file, dialect=csv.excel)\n",
    "    writer.writerow(['original_ct_value', 'metamap_input', 'metamap_input', 'metamap_cui', 'metamap_score', 'metamap_semtypes'])\n",
    "    for disease in metamapped_diseases[:100]:\n",
    "        for key in disease:\n",
    "            for concept in disease[key]:\n",
    "                concept = concept._asdict()\n",
    "                row = [key]\n",
    "                original_disease = (list(diseases_translated.keys())[list(diseases_translated.values()).index(key)])\n",
    "                row.append(original_disease)\n",
    "                row.extend([concept.get(k) for k in ['preferred_name', 'cui', 'score', 'semtypes']])\n",
    "                writer.writerow(row)\n",
    "                print(row)\n",
    "                break\n",
    "\n",
    "# with open(\"metamapped_interventions_qual_control.csv\", \"w\") as file:\n",
    "#     csv.excel.delimiter=','\n",
    "#     writer = csv.writer(file, dialect=csv.excel)\n",
    "#     writer.writerow(['input_intervention', 'metamapped_intervention_name', 'metamap_cui', 'metamap_score', 'metamap_semtypes'])\n",
    "#     for intervention in metamapped_interventions:\n",
    "#         for key in intervention:\n",
    "#             for concept in intervention[key]:\n",
    "#                 concept = concept._asdict()\n",
    "#     #             concept.get('preferred_name')\n",
    "#                 row = [key]\n",
    "#                 row.extend([concept.get(k) for k in ['preferred_name', 'cui', 'score', 'semtypes']])\n",
    "#                 writer.writerow(row)\n",
    "               \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_mapped_ct_data(metamapped_list, translation_dict):\n",
    "    callers_local_vars = inspect.currentframe().f_back.f_locals.items() # use inspect module to find the original variable name of metamapped list to use as file name (ex: metamapped_diseases gets passed as metamapped_list, but I want to retain metamapped_DISEASES as part of filename)\n",
    "    var_names = [var_name for var_name, var_val in callers_local_vars if var_val is metamapped_list]\n",
    "    with open(f\"{var_names[0]}_qual_control.csv\", \"w\") as file:\n",
    "        csv.excel.delimiter=','\n",
    "        writer = csv.writer(file, dialect=csv.excel)\n",
    "        writer.writerow(['original_ct_value', 'metamap_input', 'metamap_input', 'metamap_cui', 'metamap_score', 'metamap_semtypes'])\n",
    "        for metamapping in metamapped_list:\n",
    "            for key in metamapping:\n",
    "                print(key)\n",
    "                for concept in metamapping[key]:\n",
    "                    concept = concept._asdict()\n",
    "                    row = [key]\n",
    "                    original_concept = (list(translation_dict.keys())[list(translation_dict.values()).index(key)])\n",
    "                    row.append(original_concept)\n",
    "                    row.extend([concept.get(k) for k in ['preferred_name', 'cui', 'score', 'semtypes']])\n",
    "                    writer.writerow(row)\n",
    "                    break\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# def write_mapped_ct_data(metamapped_list, translation_dict):\n",
    "#     callers_local_vars = inspect.currentframe().f_back.f_locals.items() # use inspect module to find the original variable name of metamapped list to use as file name (ex: metamapped_diseases gets passed as metamapped_list, but I want to retain metamapped_DISEASES as part of filename)\n",
    "#     var_names = [var_name for var_name, var_val in callers_local_vars if var_val is metamapped_list]\n",
    "#     with open(f\"{var_names[0]}_qual_control.csv\", \"w\") as file:\n",
    "#         csv.excel.delimiter=','\n",
    "#         writer = csv.writer(file, dialect=csv.excel)\n",
    "#         writer.writerow(['original_ct_value', 'metamap_input', 'metamap_input', 'metamap_cui', 'metamap_score', 'metamap_semtypes'])\n",
    "# #         for disease in metamapped_diseases:\n",
    "#         for metamapping in metamapped_list:\n",
    "# #             for key in disease:\n",
    "#             for key in metamapping:\n",
    "#                 print(key)\n",
    "#                 for concept in metamapping[key]:\n",
    "#                     concept = concept._asdict()\n",
    "#                     row = [key]\n",
    "#                     original_concept = (list(translation_dict.keys())[list(translation_dict.values()).index(key)])\n",
    "#                     row.append(original_concept)\n",
    "#                     row.extend([concept.get(k) for k in ['preferred_name', 'cui', 'score', 'semtypes']])\n",
    "#                     writer.writerow(row)\n",
    "#                     break\n",
    "                \n",
    "#     with open(\"metamapped_diseases_qual_control.csv\", \"w\") as file:\n",
    "#     csv.excel.delimiter=','\n",
    "#     writer = csv.writer(file, dialect=csv.excel)\n",
    "#     writer.writerow(['original_ct_value', 'metamap_input', 'metamap_input', 'metamap_cui', 'metamap_score', 'metamap_semtypes'])\n",
    "#     for disease in metamapped_diseases[:10]:\n",
    "#         for key in disease:\n",
    "#             for concept in disease[key]:\n",
    "#                 concept = concept._asdict()\n",
    "#                 row = [key]\n",
    "#                 original_disease = (list(diseases_translated.keys())[list(diseases_translated.values()).index(key)])\n",
    "#                 row.append(original_disease)\n",
    "#                 row.extend([concept.get(k) for k in ['preferred_name', 'cui', 'score', 'semtypes']])\n",
    "#                 writer.writerow(row)\n",
    "#                 print(row)\n",
    "#                 break\n",
    "                \n",
    "\n",
    "# write_mapped_ct_data(metamapped_diseases, diseases_translated)  \n",
    "# write_mapped_ct_data(metamapped_interventions, interventions_translated)\n",
    "\n",
    "# pool.starmap(processURL, zip(urlSet, repeat(user), repeat(verboseFlag)))\n",
    "# multiprocessing.Pool.starmap(write_mapped_ct_data, zip(repeat(metamapped_diseases), repeat(diseases_translated)))\n",
    "\n",
    "with multiprocessing.Pool(multiprocessing.cpu_count() - 1) as pool:\n",
    "        pool.starmap(write_mapped_ct_data, zip(metamapped_diseases, repeat(diseases_translated)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(metamapped_diseases)\n",
    "len(metamapped_interventions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "ename": "Py4JJavaError",
     "evalue": "An error occurred while calling z:org.apache.spark.api.python.PythonRDD.runJob.\n: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 1 times, most recent failure: Lost task 0.0 in stage 0.0 (TID 0) (192.168.0.119 executor driver): org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 619, in main\n    process()\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 611, in process\n    serializer.dump_stream(out_iter, outfile)\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/serializers.py\", line 259, in dump_stream\n    vs = list(itertools.islice(iterator, batch))\n  File \"/usr/local/opt/apache-spark/libexec/python/pyspark/rdd.py\", line 1562, in takeUpToNumLeft\n    yield next(iterator)\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/util.py\", line 74, in wrapper\n    return f(*args, **kwargs)\n  File \"<ipython-input-139-b4ba8bab07b8>\", line 10, in <lambda>\nValueError: invalid literal for int() with base 10: 'original_ct_value'\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:555)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:713)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:695)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:508)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat scala.collection.Iterator.foreach(Iterator.scala:943)\n\tat scala.collection.Iterator.foreach$(Iterator.scala:943)\n\tat org.apache.spark.InterruptibleIterator.foreach(InterruptibleIterator.scala:28)\n\tat scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)\n\tat scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:105)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:49)\n\tat scala.collection.TraversableOnce.to(TraversableOnce.scala:366)\n\tat scala.collection.TraversableOnce.to$(TraversableOnce.scala:364)\n\tat org.apache.spark.InterruptibleIterator.to(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toBuffer(TraversableOnce.scala:358)\n\tat scala.collection.TraversableOnce.toBuffer$(TraversableOnce.scala:358)\n\tat org.apache.spark.InterruptibleIterator.toBuffer(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toArray(TraversableOnce.scala:345)\n\tat scala.collection.TraversableOnce.toArray$(TraversableOnce.scala:339)\n\tat org.apache.spark.InterruptibleIterator.toArray(InterruptibleIterator.scala:28)\n\tat org.apache.spark.api.python.PythonRDD$.$anonfun$runJob$1(PythonRDD.scala:166)\n\tat org.apache.spark.SparkContext.$anonfun$runJob$5(SparkContext.scala:2254)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:131)\n\tat org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1462)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base/java.lang.Thread.run(Thread.java:834)\n\nDriver stacktrace:\n\tat org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2454)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2403)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2402)\n\tat scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)\n\tat scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)\n\tat scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2402)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1160)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1160)\n\tat scala.Option.foreach(Option.scala:407)\n\tat org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1160)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2642)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2584)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2573)\n\tat org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:938)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2214)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2235)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2254)\n\tat org.apache.spark.api.python.PythonRDD$.runJob(PythonRDD.scala:166)\n\tat org.apache.spark.api.python.PythonRDD.runJob(PythonRDD.scala)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:282)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)\n\tat py4j.ClientServerConnection.run(ClientServerConnection.java:106)\n\tat java.base/java.lang.Thread.run(Thread.java:834)\nCaused by: org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 619, in main\n    process()\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 611, in process\n    serializer.dump_stream(out_iter, outfile)\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/serializers.py\", line 259, in dump_stream\n    vs = list(itertools.islice(iterator, batch))\n  File \"/usr/local/opt/apache-spark/libexec/python/pyspark/rdd.py\", line 1562, in takeUpToNumLeft\n    yield next(iterator)\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/util.py\", line 74, in wrapper\n    return f(*args, **kwargs)\n  File \"<ipython-input-139-b4ba8bab07b8>\", line 10, in <lambda>\nValueError: invalid literal for int() with base 10: 'original_ct_value'\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:555)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:713)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:695)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:508)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat scala.collection.Iterator.foreach(Iterator.scala:943)\n\tat scala.collection.Iterator.foreach$(Iterator.scala:943)\n\tat org.apache.spark.InterruptibleIterator.foreach(InterruptibleIterator.scala:28)\n\tat scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)\n\tat scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:105)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:49)\n\tat scala.collection.TraversableOnce.to(TraversableOnce.scala:366)\n\tat scala.collection.TraversableOnce.to$(TraversableOnce.scala:364)\n\tat org.apache.spark.InterruptibleIterator.to(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toBuffer(TraversableOnce.scala:358)\n\tat scala.collection.TraversableOnce.toBuffer$(TraversableOnce.scala:358)\n\tat org.apache.spark.InterruptibleIterator.toBuffer(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toArray(TraversableOnce.scala:345)\n\tat scala.collection.TraversableOnce.toArray$(TraversableOnce.scala:339)\n\tat org.apache.spark.InterruptibleIterator.toArray(InterruptibleIterator.scala:28)\n\tat org.apache.spark.api.python.PythonRDD$.$anonfun$runJob$1(PythonRDD.scala:166)\n\tat org.apache.spark.SparkContext.$anonfun$runJob$5(SparkContext.scala:2254)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:131)\n\tat org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1462)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\t... 1 more\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m----------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mPy4JJavaError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-139-b4ba8bab07b8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     13\u001b[0m                                         \u001b[0mmetamap_score\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m                                         metamap_semtypes=[4]))\n\u001b[0;32m---> 15\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mct_data_struct\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m \u001b[0;31m#convert the RDD to DataFrame\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/opt/apache-spark/libexec/python/pyspark/rdd.py\u001b[0m in \u001b[0;36mtake\u001b[0;34m(self, num)\u001b[0m\n\u001b[1;32m   1566\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1567\u001b[0m             \u001b[0mp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpartsScanned\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpartsScanned\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mnumPartsToTry\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtotalParts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1568\u001b[0;31m             \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontext\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrunJob\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtakeUpToNumLeft\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1569\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1570\u001b[0m             \u001b[0mitems\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/opt/apache-spark/libexec/python/pyspark/context.py\u001b[0m in \u001b[0;36mrunJob\u001b[0;34m(self, rdd, partitionFunc, partitions, allowLocal)\u001b[0m\n\u001b[1;32m   1225\u001b[0m         \u001b[0;31m# SparkContext#runJob.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1226\u001b[0m         \u001b[0mmappedRDD\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmapPartitions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpartitionFunc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1227\u001b[0;31m         \u001b[0msock_info\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jvm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPythonRDD\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrunJob\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jsc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmappedRDD\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jrdd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpartitions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1228\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_load_from_socket\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msock_info\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmappedRDD\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jrdd_deserializer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1229\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/opt/apache-spark/libexec/python/lib/py4j-0.10.9.3-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m   1320\u001b[0m         \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgateway_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1321\u001b[0m         return_value = get_return_value(\n\u001b[0;32m-> 1322\u001b[0;31m             answer, self.gateway_client, self.target_id, self.name)\n\u001b[0m\u001b[1;32m   1323\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1324\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mtemp_arg\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtemp_args\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/opt/apache-spark/libexec/python/pyspark/sql/utils.py\u001b[0m in \u001b[0;36mdeco\u001b[0;34m(*a, **kw)\u001b[0m\n\u001b[1;32m    109\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mdeco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    110\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 111\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    112\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mpy4j\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotocol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPy4JJavaError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    113\u001b[0m             \u001b[0mconverted\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconvert_exception\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjava_exception\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/opt/apache-spark/libexec/python/lib/py4j-0.10.9.3-src.zip/py4j/protocol.py\u001b[0m in \u001b[0;36mget_return_value\u001b[0;34m(answer, gateway_client, target_id, name)\u001b[0m\n\u001b[1;32m    326\u001b[0m                 raise Py4JJavaError(\n\u001b[1;32m    327\u001b[0m                     \u001b[0;34m\"An error occurred while calling {0}{1}{2}.\\n\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 328\u001b[0;31m                     format(target_id, \".\", name), value)\n\u001b[0m\u001b[1;32m    329\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    330\u001b[0m                 raise Py4JError(\n",
      "\u001b[0;31mPy4JJavaError\u001b[0m: An error occurred while calling z:org.apache.spark.api.python.PythonRDD.runJob.\n: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 1 times, most recent failure: Lost task 0.0 in stage 0.0 (TID 0) (192.168.0.119 executor driver): org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 619, in main\n    process()\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 611, in process\n    serializer.dump_stream(out_iter, outfile)\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/serializers.py\", line 259, in dump_stream\n    vs = list(itertools.islice(iterator, batch))\n  File \"/usr/local/opt/apache-spark/libexec/python/pyspark/rdd.py\", line 1562, in takeUpToNumLeft\n    yield next(iterator)\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/util.py\", line 74, in wrapper\n    return f(*args, **kwargs)\n  File \"<ipython-input-139-b4ba8bab07b8>\", line 10, in <lambda>\nValueError: invalid literal for int() with base 10: 'original_ct_value'\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:555)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:713)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:695)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:508)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat scala.collection.Iterator.foreach(Iterator.scala:943)\n\tat scala.collection.Iterator.foreach$(Iterator.scala:943)\n\tat org.apache.spark.InterruptibleIterator.foreach(InterruptibleIterator.scala:28)\n\tat scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)\n\tat scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:105)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:49)\n\tat scala.collection.TraversableOnce.to(TraversableOnce.scala:366)\n\tat scala.collection.TraversableOnce.to$(TraversableOnce.scala:364)\n\tat org.apache.spark.InterruptibleIterator.to(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toBuffer(TraversableOnce.scala:358)\n\tat scala.collection.TraversableOnce.toBuffer$(TraversableOnce.scala:358)\n\tat org.apache.spark.InterruptibleIterator.toBuffer(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toArray(TraversableOnce.scala:345)\n\tat scala.collection.TraversableOnce.toArray$(TraversableOnce.scala:339)\n\tat org.apache.spark.InterruptibleIterator.toArray(InterruptibleIterator.scala:28)\n\tat org.apache.spark.api.python.PythonRDD$.$anonfun$runJob$1(PythonRDD.scala:166)\n\tat org.apache.spark.SparkContext.$anonfun$runJob$5(SparkContext.scala:2254)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:131)\n\tat org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1462)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base/java.lang.Thread.run(Thread.java:834)\n\nDriver stacktrace:\n\tat org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2454)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2403)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2402)\n\tat scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)\n\tat scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)\n\tat scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2402)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1160)\n\tat org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1160)\n\tat scala.Option.foreach(Option.scala:407)\n\tat org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1160)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2642)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2584)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2573)\n\tat org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:938)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2214)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2235)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2254)\n\tat org.apache.spark.api.python.PythonRDD$.runJob(PythonRDD.scala:166)\n\tat org.apache.spark.api.python.PythonRDD.runJob(PythonRDD.scala)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:282)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)\n\tat py4j.ClientServerConnection.run(ClientServerConnection.java:106)\n\tat java.base/java.lang.Thread.run(Thread.java:834)\nCaused by: org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 619, in main\n    process()\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/worker.py\", line 611, in process\n    serializer.dump_stream(out_iter, outfile)\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/serializers.py\", line 259, in dump_stream\n    vs = list(itertools.islice(iterator, batch))\n  File \"/usr/local/opt/apache-spark/libexec/python/pyspark/rdd.py\", line 1562, in takeUpToNumLeft\n    yield next(iterator)\n  File \"/usr/local/opt/apache-spark/libexec/python/lib/pyspark.zip/pyspark/util.py\", line 74, in wrapper\n    return f(*args, **kwargs)\n  File \"<ipython-input-139-b4ba8bab07b8>\", line 10, in <lambda>\nValueError: invalid literal for int() with base 10: 'original_ct_value'\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:555)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:713)\n\tat org.apache.spark.api.python.PythonRunner$$anon$3.read(PythonRunner.scala:695)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:508)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat scala.collection.Iterator.foreach(Iterator.scala:943)\n\tat scala.collection.Iterator.foreach$(Iterator.scala:943)\n\tat org.apache.spark.InterruptibleIterator.foreach(InterruptibleIterator.scala:28)\n\tat scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)\n\tat scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:105)\n\tat scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:49)\n\tat scala.collection.TraversableOnce.to(TraversableOnce.scala:366)\n\tat scala.collection.TraversableOnce.to$(TraversableOnce.scala:364)\n\tat org.apache.spark.InterruptibleIterator.to(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toBuffer(TraversableOnce.scala:358)\n\tat scala.collection.TraversableOnce.toBuffer$(TraversableOnce.scala:358)\n\tat org.apache.spark.InterruptibleIterator.toBuffer(InterruptibleIterator.scala:28)\n\tat scala.collection.TraversableOnce.toArray(TraversableOnce.scala:345)\n\tat scala.collection.TraversableOnce.toArray$(TraversableOnce.scala:339)\n\tat org.apache.spark.InterruptibleIterator.toArray(InterruptibleIterator.scala:28)\n\tat org.apache.spark.api.python.PythonRDD$.$anonfun$runJob$1(PythonRDD.scala:166)\n\tat org.apache.spark.SparkContext.$anonfun$runJob$5(SparkContext.scala:2254)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:131)\n\tat org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1462)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\t... 1 more\n"
     ]
    }
   ],
   "source": [
    "from pyspark.sql import Row\n",
    "# spark - sparkSession\n",
    "# sc = spark.sparkContext\n",
    "\n",
    "# Load a text file and convert each line to a Row.\n",
    "ct_data = sc.textFile(\"metamapped_diseases_qual_control.csv\")\n",
    "#Split on delimiters\n",
    "ct_items = ct_data.map(lambda l: l.split(\",\"))\n",
    "#Convert to Row\n",
    "ct_data_struct = ct_items.map(lambda p: Row(original_ct_value=int(p[0]),\n",
    "                                        metamap_input=p[1],\n",
    "                                        metamap_cui=p[2],\n",
    "                                        metamap_score=p[3],\n",
    "                                        metamap_semtypes=[4]))\n",
    "for i in ct_data_struct.take(5): print(i)\n",
    "#convert the RDD to DataFrame\n",
    "\n",
    "ct_df = spark.createDataFrame(ct_data_struct)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'array' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m----------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-143-24e6fc2c992e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m schema = StructType(\n\u001b[0;32m----> 5\u001b[0;31m     array(\n\u001b[0m\u001b[1;32m      6\u001b[0m         \u001b[0mStructField\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"original_ct_value\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mStringType\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mtrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m         \u001b[0mStructField\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"metamap_input\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mStringType\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mtrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'array' is not defined"
     ]
    }
   ],
   "source": [
    "from pyspark.sql.types import StructType, StructField\n",
    "from pyspark.sql.types import StringType, ArrayType\n",
    "\n",
    "schema = StructType(\n",
    "    Array(\n",
    "        StructField(\"original_ct_value\", StringType,true),\n",
    "        StructField(\"metamap_input\", StringType,true),\n",
    "        StructField(\"metamap_output_preferred_name\", StringType,true),\n",
    "        StructField(\"cui\", StringType,true),\n",
    "        StructField(\"score\", StringType,true),\n",
    "        StructField(\"semtypes\", StringType,true),\n",
    "        \n",
    "))\n",
    "\n",
    "test = spark.read.csv(\n",
    "    \"metamapped_diseases_qual_control.csv\", \n",
    "    header=True, \n",
    "    mode=\"DROPMALFORMED\", \n",
    "    schema=schema\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'sc'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m----------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-136-f6bb7c0d74e6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# Load a text file and convert each line to a Row.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0morders\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtextFile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"/practicedata/orders\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0;31m#Split on delimiters\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mparts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0morders\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0ml\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0ml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\",\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/opt/apache-spark/libexec/python/pyspark/context.py\u001b[0m in \u001b[0;36mtextFile\u001b[0;34m(self, name, minPartitions, use_unicode)\u001b[0m\n\u001b[1;32m    652\u001b[0m         \u001b[0;34m[\u001b[0m\u001b[0;34m'Hello world!'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    653\u001b[0m         \"\"\"\n\u001b[0;32m--> 654\u001b[0;31m         \u001b[0mminPartitions\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mminPartitions\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdefaultParallelism\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    655\u001b[0m         return RDD(self._jsc.textFile(name, minPartitions), self,\n\u001b[1;32m    656\u001b[0m                    UTF8Deserializer(use_unicode))\n",
      "\u001b[0;32m/usr/local/opt/apache-spark/libexec/python/pyspark/context.py\u001b[0m in \u001b[0;36mdefaultParallelism\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    448\u001b[0m         reduce tasks)\n\u001b[1;32m    449\u001b[0m         \"\"\"\n\u001b[0;32m--> 450\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jsc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdefaultParallelism\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    451\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    452\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'sc'"
     ]
    }
   ],
   "source": [
    "from pyspark.sql import Row\n",
    "# spark - sparkSession\n",
    "sc = spark.sparkContext\n",
    "\n",
    "# Load a text file and convert each line to a Row.\n",
    "orders = sc.textFile(\"/practicedata/orders\")\n",
    "#Split on delimiters\n",
    "parts = orders.map(lambda l: l.split(\",\"))\n",
    "#Convert to Row\n",
    "orders_struct = parts.map(lambda p: Row(order_id=int(p[0]), order_date=p[1], customer_id=p[2], order_status=p[3]))\n",
    "for i in orders_struct.take(5): print(i)\n",
    "#convert the RDD to DataFrame\n",
    "\n",
    "orders_df = spark.createDataFrame(orders_struct)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pyspark\n",
    "from pyspark.sql import Row\n",
    "\n",
    "spark = SparkSession.builder.master(\"local[2]\") \\\n",
    "                    .appName('clinical_trials_etl') \\\n",
    "                    .getOrCreate()\n",
    "\n",
    "# # test = spark.createDataFrame(Row(**x) for x in metamapped_diseases[:20]).show(truncate=False)\n",
    "# for disease in metamapped_diseases:\n",
    "#         for key in disease:\n",
    "#             for concept in disease[key]:\n",
    "#                 concept = concept._asdict()\n",
    "#                 row = [key]\n",
    "#                 row.extend([concept.get(k) for k in ['preferred_name', 'cui', 'score', 'semtypes']])\n",
    "#                 writer.writerow(row)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'infertile high-risk patients with polycystic ovaries': [ConceptMMI(index='USER', mm='MMI', score='22.36', preferred_name='Polycystic Ovary Syndrome', cui='C0032460', semtypes='[dsyn]', trigger='[\"Polycystic ovary syndrome\"-tx-1-\"polycystic ovaries\"-noun-0]', location='TX', pos_info='38/18', tree_codes='C04.182.612.765;C13.351.500.056.630.580.765;C19.391.630.580.765'),\n",
       "   ConceptMMI(index='USER', mm='MMI', score='12.88', preferred_name='Infertility', cui='C0021359', semtypes='[patf]', trigger='[\"Infertility\"-tx-1-\"infertile\"-adj-0]', location='TX', pos_info='4/9', tree_codes='C12.294.365;C13.351.500.365'),\n",
       "   ConceptMMI(index='USER', mm='MMI', score='6.70', preferred_name='Patients', cui='C0030705', semtypes='[podg]', trigger='[\"Patient\"-tx-1-\"patients\"-noun-0]', location='TX', pos_info='24/8', tree_codes='M01.643'),\n",
       "   ConceptMMI(index='USER', mm='MMI', score='3.57', preferred_name='High risk', cui='C4319571', semtypes='[fndg]', trigger='[\"High risk\"-tx-1-\"high-risk\"-noun-0]', location='TX', pos_info='14/9', tree_codes=''),\n",
       "   ConceptMMI(index='USER', mm='MMI', score='3.57', preferred_name='High risk of', cui='C0332167', semtypes='[fndg]', trigger='[\"High risk of\"-tx-1-\"high-risk\"-noun-0]', location='TX', pos_info='14/9', tree_codes='')]},\n",
       " {'anaesthetic induction': [ConceptMMI(index='USER', mm='MMI', score='19.25', preferred_name='Anesthetics', cui='C0002932', semtypes='[phsu]', trigger='[\"Anesthetic, NOS\"-tx-1-\"anaesthetic\"-adj-0]', location='TX', pos_info='4/11', tree_codes='D27.505.696.277.100;D27.505.954.427.210.100'),\n",
       "   ConceptMMI(index='USER', mm='MMI', score='3.63', preferred_name='Induce (action)', cui='C0205263', semtypes='[ftcn]', trigger='[\"Induced\"-tx-1-\"induction\"-noun-0]', location='TX', pos_info='16/9', tree_codes='')]}]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metamapped_diseases[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'infertile high-risk patients with polycystic ovaries': [ConceptMMI(index='USER', mm='MMI', score='22.36', preferred_name='Polycystic Ovary Syndrome', cui='C0032460', semtypes='[dsyn]', trigger='[\"Polycystic ovary syndrome\"-tx-1-\"polycystic ovaries\"-noun-0]', location='TX', pos_info='38/18', tree_codes='C04.182.612.765;C13.351.500.056.630.580.765;C19.391.630.580.765'),\n",
       "   ConceptMMI(index='USER', mm='MMI', score='12.88', preferred_name='Infertility', cui='C0021359', semtypes='[patf]', trigger='[\"Infertility\"-tx-1-\"infertile\"-adj-0]', location='TX', pos_info='4/9', tree_codes='C12.294.365;C13.351.500.365'),\n",
       "   ConceptMMI(index='USER', mm='MMI', score='6.70', preferred_name='Patients', cui='C0030705', semtypes='[podg]', trigger='[\"Patient\"-tx-1-\"patients\"-noun-0]', location='TX', pos_info='24/8', tree_codes='M01.643'),\n",
       "   ConceptMMI(index='USER', mm='MMI', score='3.57', preferred_name='High risk', cui='C4319571', semtypes='[fndg]', trigger='[\"High risk\"-tx-1-\"high-risk\"-noun-0]', location='TX', pos_info='14/9', tree_codes=''),\n",
       "   ConceptMMI(index='USER', mm='MMI', score='3.57', preferred_name='High risk of', cui='C0332167', semtypes='[fndg]', trigger='[\"High risk of\"-tx-1-\"high-risk\"-noun-0]', location='TX', pos_info='14/9', tree_codes='')]},\n",
       " {'anaesthetic induction': [ConceptMMI(index='USER', mm='MMI', score='19.25', preferred_name='Anesthetics', cui='C0002932', semtypes='[phsu]', trigger='[\"Anesthetic, NOS\"-tx-1-\"anaesthetic\"-adj-0]', location='TX', pos_info='4/11', tree_codes='D27.505.696.277.100;D27.505.954.427.210.100'),\n",
       "   ConceptMMI(index='USER', mm='MMI', score='3.63', preferred_name='Induce (action)', cui='C0205263', semtypes='[ftcn]', trigger='[\"Induced\"-tx-1-\"induction\"-noun-0]', location='TX', pos_info='16/9', tree_codes='')]}]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pyspark.sql import Row\n",
    "spark.createDataFrame(Row(**x) for x in mylist).show(truncate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# pip show pyspark\n",
    "\n",
    "\n",
    "# import pyspark\n",
    "# from pyspark.sql import SparkSession\n",
    "# spark = SparkSession.builder.master(\"local[1]\").appName(\"SparkByExamples.com\").getOrCreate()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "spark.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# spark = SparkSession.builder.appName(\"test\").config(\n",
    "#     \"spark.driver.extraJavaOptions\",\n",
    "#     \"--add-opens=java.base/sun.nio.ch=ALL-UNNAMED --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED\",\n",
    "# ).getOrCreate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# link nct ids of studies that have the same disease and intervention pair---multiple studies may have studies the same intervention\n",
    "# restrict returned CUIs of metamapped concepts???"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "user = os.environ['USER']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Kamileh'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "user"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b'java version \"1.8.0_251\"\\nJava(TM) SE Runtime Environment (build 1.8.0_251-b08)\\nJava HotSpot(TM) 64-Bit Server VM (build 25.251-b08, mixed mode)\\n'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "version = subprocess.check_output(['java', '-version'], stderr=subprocess.STDOUT)\n",
    "version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.width', None)\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "mapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ct_extract_env",
   "language": "python",
   "name": "ct_extract_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
